---
title: "Testes Enrichment Analyses"
---
# Prep
## Specify paths
```{r parameters}
#All input files on the farm
path <- "/Users/mn7/volumes/mn7_lustre/pipelineInput/"
#Local output
plotsPath <- "/Users/mn7/Google Drive File Stream/My Drive/Testes/Signature_Enrichment/plots/"
#Farm output
#plotsPath <- "/Users/mn7/volumes/mn7_lustre/pipelineInput/plots/testisIndivJun26_20/"

```

## Load Libraries
```{r libraries}
library(reshape2)
library(Rsamtools)
library(Repitools)
library(plyr)
library(sigfit)
library(devtools)
library(patchwork)
library(MutationalPatterns)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(ref_genome, character.only = TRUE)
library(lme4)
library(broom)
library(tidyverse)
library(dndscv)
library(gtools)
select <- dplyr::select
rename <- dplyr::rename
summarize <- dplyr::summarize
slice <- dplyr::slice
options(scipen = 999)

```


## Generate MutPattern Input Regions

*** Set to eval = FALSE as these are time consuming to run and only necessary if you do not have the .RData objects they produce ***
*** Save lines commented out to prevent accidental overwrite ***

```{r mutPatternRegions, eval=FALSE}
# Misc regions
misc_regions_files = list.files(paste0(path,"mutPatternsRegions/genomic_regions"), full.names = T)
print(misc_regions_files)
misc_regions_files = misc_regions_files[c(1,2,6,7)]

# region bed files
misc_regions_list=list()
for(i in misc_regions_files){
  misc_region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(misc_region) <- "UCSC"
  misc_regions_list[[i]] <- misc_region
}

names(misc_regions_list) = c("H3k27ac", "H3K9me3", "Autosomal_exon", "Coding")
#save(misc_regions_list, file=paste0(path,"mutPatternsRegions/misc_regions_list.RData"))

#Alt reptiming
H9es1mb <- read_tsv(paste0(path, "mutPatternsRegions/repTimeH9es.1mb.bed"), col_names = c("chr", "start", "end", "repTime"), col_types = 'ciic') %>% 
  mutate(repTime = if_else(repTime == ".", "NA", repTime)) %>% 
  mutate(repTime = as.numeric(repTime))
hist(H9es1mb$repTime)

LCL1mb <- read_tsv(paste0(path, "mutPatternsRegions/repTimeLCL.1mb.bed"), col_names = c("chr", "start", "end", "repTime"), col_types = 'ciic') %>% 
  mutate(repTime = if_else(repTime == ".", "NA", repTime)) %>% 
  mutate(repTime = as.numeric(repTime))
hist(H9es1mb$repTime)

#Split into 3 roughly equal groups
write_tsv(H9es1mb %>% filter(repTime > 0.5), path = paste0(path, "mutPatternsRegions/H9repTiming/H9early.bed"), col_names = F)
write_tsv(H9es1mb %>% filter(repTime > -0.5 & repTime < 0.5), path = paste0(path, "mutPatternsRegions/H9repTiming/H9intermediate.bed"), col_names = F)
write_tsv(H9es1mb %>% filter(repTime < -0.5), path = paste0(path, "mutPatternsRegions/H9repTiming/H9late.bed"), col_names = F)

write_tsv(LCL1mb %>% filter(repTime > 0.5), path = paste0(path, "mutPatternsRegions/LCLrepTiming/LCLearly.bed"), col_names = F)
write_tsv(LCL1mb %>% filter(repTime > -0.5 & repTime < 0.5), path = paste0(path, "mutPatternsRegions/LCLrepTiming/LCLintermediate.bed"), col_names = F)
write_tsv(LCL1mb %>% filter(repTime < -0.5), path = paste0(path, "mutPatternsRegions/LCLrepTiming/LCLlate.bed"), col_names = F)
```


# Functions
Functions to generate variants and callable regions for each dataset

## Load Variant VCFs
```{r vcfLoadFunction}

load_VCF <- function(dataset) {
  #Pop variants from gnomAD
  if(dataset == "PopVars") {
    gnomadVCF <- read_vcfs_as_granges(paste0(path,"variantVCFs/PopVars/gnomad.genomes.r2.1.1.snps.filtered.vcf"), "gnomad", ref_genome, check_alleles=T)
    return(gnomadVCF)
  }
  #Testes by individual
  if(dataset == "Testes") {
    testesIndivFiles <- list.files(paste0(path, "variantVCFs/TestesNormal"), pattern = ".vcf.gz", full.names = TRUE)
    testesIndivSamples <-  str_remove(list.files(paste0(path, "variantVCFs/TestesNormal"), pattern = ".vcf.gz"), ".vcf.gz")
    testesIndivVCF <- read_vcfs_as_granges(testesIndivFiles, testesIndivSamples, ref_genome, check_alleles=T)
    return(testesIndivVCF)
  }
  #Benign Testes by individual
  if(dataset == "TestesBenign") {
    benignIndivFiles <- list.files(paste0(path, "variantVCFs/TestesBenign"), pattern = ".vcf.gz", full.names = TRUE)
    benignIndivSamples <-  str_remove(list.files(paste0(path, "variantVCFs/TestesBenign"), pattern = ".vcf.gz"), ".vcf.gz")
    benignIndivVCF <- read_vcfs_as_granges(benignIndivFiles, benignIndivSamples, ref_genome, check_alleles=T)
    return(benignIndivVCF)
  }
  #Colon by individual
  if(dataset == "Colon") {
    colonIndivFiles <- list.files(paste0(path, "variantVCFs/Colon"), pattern = ".vcf.gz", full.names = TRUE)
    colonIndivSamples <-  str_remove(list.files(paste0(path, "variantVCFs/Colon"), pattern = ".vcf.gz"), ".vcf.gz")
    colonIndivVCF <- read_vcfs_as_granges(colonIndivFiles, colonIndivSamples, ref_genome, check_alleles=T)
    return(colonIndivVCF)
  }
  #Trios from DeCode
  if(dataset == "Trios") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/trios_decode.vcf"), "Trio_data", ref_genome, check_alleles=T)
    return(triosVCF)
  }
  #Trios paternally phased
  if(dataset == "TriosPat") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/triosPaternal_hg19.vcf"), "TriosPat", ref_genome, check_alleles=T)
    return(triosVCF)
  }
  #Trios maternally phased
  if(dataset == "TriosMat") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/triosMaternal_hg19.vcf"), "TriosMat", ref_genome, check_alleles=T)
    return(triosVCF)
  }  
  #Lung Never-Smoker
  if(dataset == "LungNormal") {
    lungNormalVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/LungNormal/Neversmoker_data.unique.vcf"), "LungNormal", ref_genome, check_alleles=T)
    return(lungNormalVCF)
  }
  #Lung Ex-Smoker
  if(dataset == "LungExSmoker") {
    lungExSmokerVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/LungExSmoker/Ex_smokers_data.unique.vcf"), "LungExSmoker", ref_genome, check_alleles=T)
    return(lungExSmokerVCF)
  }
  #Sperm Normal
  if(dataset == "Sperm") {
    healthySpermFiles <- list.files(paste0(path, "variantVCFs/Sperm/"), pattern = ".vcf", full.names = TRUE)
    healthySpermSamples <- c("sperm15", "sperm16")
    healthySpermVCF <- read_vcfs_as_granges(healthySpermFiles, healthySpermSamples, ref_genome, check_alleles=T)
    return(healthySpermVCF)
  }
  #POL mutatnt Sperm/Blood
  if(dataset == "POL") {
    botseqIndivFiles <- list.files(paste0(path, "variantVCFs/POL"), pattern = ".vcf", full.names = TRUE)
    botseqIndivSamples <- c("POLE_Blood", "POLD1_Blood", "POLE_Sperm", "POLD1_Sperm")
    botseqIndivVCF <- read_vcfs_as_granges(botseqIndivFiles, botseqIndivSamples, ref_genome, check_alleles=T)
    return(botseqIndivVCF)
  }
  #PanBody muts
  if(dataset == "PanBody") {
    panBodyFiles <- list.files(paste0(path, "variantVCFs/PanBody"), pattern = ".vcf.gz" , full.names = TRUE)
    panBodySamples <- str_remove(list.files(paste0(path, "variantVCFs/PanBody"), pattern = ".vcf.gz"), ".vcf.gz")
    panBodyVCF <- read_vcfs_as_granges(panBodyFiles, panBodySamples, ref_genome, check_alleles=T)
    return(panBodyVCF)
  }
  #Germline Tumours
  if(dataset == "Tolly_Tumour") {
    tumourIndivFiles <- list.files(paste0(path, "variantVCFs/Tolly_Tumour"), pattern = ".vcf.gz", full.names = TRUE)
    tumourIndivSamples <- str_remove(list.files(paste0(path, "variantVCFs/Tolly_Tumour"), pattern = ".vcf.gz"), ".vcf.gz")
    tumourIndivVCF <- read_vcfs_as_granges(tumourIndivFiles ,tumourIndivSamples, ref_genome, check_alleles=T)
    return(tumourIndivVCF)
  }
}

```

## Load Surveyed Regions
*** Set to preloaded = T as these are time consuming to run and only necessary if you do not have the .RData objects as shortcuts ***
```{r surveyedFunction}

load_callable <- function(dataset, preloaded = T) {
  #Full genome for hg19, no adjustment for callable regions
  if(dataset == "generic") {
    full_surveyed <- import(paste0(path, "surveyedRegions/full.callable.bed"))
    seqlevelsStyle(full_surveyed) <- "UCSC"
    return(full_surveyed)
  }
  #gnomad callable regions
  if(dataset == "PopVars") {
    pop_surveyed <- import(paste0(path, "surveyedRegions/PopVars/gnomadGenomesCallable.2.1.1.bed"))
    seqlevelsStyle(pop_surveyed) <- "UCSC"
    return(pop_surveyed)
  }
  #Testes: Genome coverage calculated by merging callable region bed files of each individual's samples
  if(dataset == "Testes") {
    if(preloaded) {
      load(paste0(path, "surveyedRegions/TestesNormal/testesIndivSurveyedList.RData"))
      return(testesIndivSurveyedList)
    } 
    else {
      testesIndivSamples <- list.files(paste0(path, "surveyedRegions/TestesNormal/"), pattern = ".sorted.merged.bed")
      testesIndivSurveyedList <- vector(mode = "list")
      for(i in testesIndivSamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/TestesNormal/",i))
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        testesIndivSurveyedList[[i]] <- surveyed
      }
      save(testesIndivSurveyedList, file= paste0(path,"surveyedRegions/TestesNormal/testesIndivSurveyedList.RData"))
      return(testesIndivSurveyedList)
    }
  }
  #Testes Benign: Genome coverage calculated by merging callable region bed files of each individual's samples
  if(dataset == "TestesBenign") {
    if(preloaded) {
      load(paste0(path, "surveyedRegions/TestesBenign/testesBenignSurveyedList.RData"))
      return(testesBenignSurveyedList)
    } 
    else {
      testesBenignSamples <- list.files(paste0(path, "surveyedRegions/TestesBenign/"), pattern = ".sorted.merged.bed")
      testesBenignSurveyedList <- vector(mode = "list")
      for(i in testesBenignSamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/TestesBenign/",i))
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        testesBenignSurveyedList[[i]] <- surveyed
      }
      save(testesBenignSurveyedList, file= paste0(path,"surveyedRegions/TestesBenign/testesBenignSurveyedList.RData"))
      return(testesBenignSurveyedList)
    }
  }
  #Colon: Genome coverage calculated by merging callable region bed files of each individual's samples
  if(dataset == "Colon") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/Colon/colonIndivSurveyedList.RData"))
      return(colonIndivSurveyedList)
    } 
    else {
      colonIndivSamples <- list.files(paste0(path, "surveyedRegions/Colon/"), pattern = ".sorted.merged.bed")
      colonIndivSurveyedList <- vector(mode = "list")
      for(i in colonIndivSamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/Colon/",i))
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        colonIndivSurveyedList[[i]] <- surveyed
      }
      save(colonIndivSurveyedList, file=paste0(path,"surveyedRegions/Colon/colonIndivSurveyedList.RData"))
      return(colonIndivSurveyedList)
    }
  }
  #Healthy Sperm: Using botseq callable
  if(dataset == "Sperm") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/Sperm/healthySpermSurveyedList.RData"))
      return(healthySpermSurveyedList)
    } 
    else {
      spermSurveyed1 <- import(paste0(path,"/surveyedRegions/Sperm/genome.coverage.15.segments.bed"))
      spermSurveyed2 <- import(paste0(path,"/surveyedRegions/Sperm/genome.coverage.16.segments.bed"))
      seqlevelsStyle(spermSurveyed1) <- "UCSC"
      seqlevelsStyle(spermSurveyed2) <- "UCSC"
      healthySpermSurveyedList <- vector(mode = "list")
      healthySpermSurveyedList[[1]] <- surveyed1
      healthySpermSurveyedList[[2]] <- surveyed2
      save(healthySpermSurveyedList, file=paste0(path,"surveyedRegions/Sperm/healthySpermSurveyedList.RData"))
      return(healthySpermSurveyedList)
    }
  }
  #POL mutants: Using botseq callable
  if(dataset == "POL") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/POL/botseqPOLsurveyedList.RData"))
      return(botseqPOLsurveyedList)
    } 
    else {
      POLsurveyed1 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.20.segments.bed"))
      POLsurveyed2 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.21.segments.bed"))
      POLsurveyed3 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.22.segments.bed"))
      POLsurveyed4 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.23.segments.bed"))
      seqlevelsStyle(surveyed3) <- "UCSC"
      seqlevelsStyle(surveyed4) <- "UCSC"
      seqlevelsStyle(surveyed5) <- "UCSC"
      seqlevelsStyle(surveyed6) <- "UCSC"
      botseqPOLsurveyedList <- vector(mode = "list")
      botseqPOLsurveyedList[[1]] <- POLsurveyed1 
      botseqPOLsurveyedList[[2]] <- POLsurveyed2 
      botseqPOLsurveyedList[[3]] <- POLsurveyed3 
      botseqPOLsurveyedList[[4]] <- POLsurveyed4 
      save(botseqPOLsurveyedList, file=paste0(path,"surveyedRegions/POL/botseqPOLsurveyedList.RData"))
      return(botseqPOLsurveyedList)
    }
  }
  #PanBody: Callable regions per sample
  if(dataset == "PanBody") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/PanBody/PanBodySurveyedList.RData"))
      return(PanBodySurveyedList)
    } 
    else {
      PanBodySamples <- list.files(paste0(path, "surveyedRegions/PanBody/"), pattern = ".sorted.merged.bed")
      PanBodySurveyedList <- vector(mode = "list")
      for(i in PanBodySamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/PanBody/",i))
        surveyed <- keepStandardChromosomes(surveyed, pruning.mode = "coarse")
        surveyed <- dropSeqlevels(surveyed, "chrM", pruning.mode="coarse")
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        PanBodySurveyedList[[i]] <- surveyed
      }
      save(PanBodySurveyedList, file=paste0(path,"surveyedRegions/PanBody/PanBodySurveyedList.RData"))
      return(PanBodySurveyedList)
    }
  }
}
```

## Load TriRate List
```{r triRateCalc}

loadTriRate <- function(vcf_list, surveyed_list, tissues) {
  triRateDF <- tibble()
  for(tissue in unique(tissues)) {
    #Subset vcfs by tissue
    vcfsT <- vcf_list[tissues == tissue]
    #Get total muts for tissue
    muts0 <- mut_matrix(vcfsT, ref_genome)
    muts <- as_tibble(muts0) %>% 
      mutate(mutType = rownames(muts0)) %>% 
      rowwise() %>%
      mutate(totMuts = sum(c_across(where(is.numeric)))) %>% 
      mutate(context = paste0(str_sub(mutType,1,1),str_sub(mutType,3,3),str_sub(mutType,7,7))) %>% 
      group_by(context) %>% 
      summarise(count = sum(totMuts), .groups = "drop")
      
    #Get mean opportunities for tissue
    surveyedT <- surveyed_list[tissues == tissue]
    for(i in 1:length(surveyedT)) {
      sequences <- getSeq(BSgenome.Hsapiens.UCSC.hg19, surveyedT[[i]])
      #Count trinuc freq
      tri_freqs <- trinucleotideFrequency(sequences)
      complement = vector()
      complement["A"] = "T"; complement["C"] = "G"; complement["G"] = "C"; complement["T"] = "A";	
      tri_freqs_summary <- as.tibble(tri_freqs) %>% 
        summarise(across(.cols = everything(), sum)) %>% 
        pivot_longer(cols = everything(), names_to = "context0", values_to = "count") %>% 
        #Convert to pyr reference
        mutate(context = ifelse(str_sub(context0,2,2) %in% c("C", "T"), context0, paste0(complement[str_sub(context0, 3,3)], complement[str_sub(context0, 2,2)], complement[str_sub(context0, 1,1)]))) %>% 
        group_by(context) %>% 
        summarise(count = sum(count), .groups = "drop")
      colnames(tri_freqs_summary) <- c("context", paste0("opps_", names(surveyedT[i])))
      muts <- muts %>% 
        left_join(tri_freqs_summary, by = "context")
    }
    freqDF <- muts %>% 
      rowwise() %>%
      mutate(meanOpps = mean(c_across(cols = starts_with("opps_")))) %>% 
      mutate(rate = count/meanOpps) %>% 
      mutate(rate = rate/sum(lengths(vcfsT))) %>% 
      mutate(tissue = tissue) %>% 
      select(context, tissue, rate)
    triRateDF <- triRateDF %>% 
      bind_rows(freqDF)
  }
  return(triRateDF)
}
```

## Custom Enrich Fxns
```{r customEnrichFxn}
#Edit intersect_with_region and genomic_distribution fxns from mutational patterns 
#to calculate expected value while controlling for triNuc mutation rate of the tissue type
intersect_with_regionCustom = function(vcf, surveyed, region, triRateTable, tissueType) {
    # Number of mutations in vcf file
    n_muts = length(vcf)
    # Number of base pairs that were surveyed
    surveyed_length = sum(as.numeric(width(surveyed)))
    # Check if chromosome names are the same in the objects
    if (seqlevelsStyle(vcf) != seqlevelsStyle(surveyed))
        stop(paste("The chromosome names (seqlevels) of the VCF and the",
                    "surveyed GRanges object do not match."))
    if (seqlevelsStyle(region) != seqlevelsStyle(surveyed))
        stop(paste("The chromosome names (seqlevels) of the surveyed and",
                    "the region GRanges object do not match."))
    # Intersect genomic region and surveyed region
    surveyed_region = intersect(surveyed, region, ignore.strand = TRUE)
    surveyed_region_length = sum(width(surveyed_region))
    #Add base to start and end to get correct trinuc counts
    surveyed_region@ranges@start <- as.integer(surveyed_region@ranges@start - 1)
    surveyed_region@ranges@width <- as.integer(surveyed_region@ranges@width + 2)
    #Count trinuc freq
    cat("Extracting sequences...\n")
  	sequences = getSeq(BSgenome.Hsapiens.UCSC.hg19, surveyed_region)
    cat("Calculating trinucleotide frequencies...\n")
    tri_freqs <- trinucleotideFrequency(sequences)
    
    complement = vector()
    complement["A"] = "T"; complement["C"] = "G"; complement["G"] = "C"; complement["T"] = "A";	
    tri_freqs_summary <- as.tibble(tri_freqs) %>% 
      summarise(across(.cols = everything(), sum)) %>% 
      pivot_longer(cols = everything(), names_to = "context0", values_to = "count") %>% 
      #Convert to pyr reference
      mutate(context = ifelse(str_sub(context0,2,2) %in% c("C", "T"), context0, paste0(complement[str_sub(context0, 3,3)], complement[str_sub(context0, 2,2)], complement[str_sub(context0, 1,1)]))) %>% 
      group_by(context) %>% 
      summarise(count = sum(count), .groups = "drop") %>% 
      left_join((triRateTable %>% filter(tissue == tissueType)), by = "context") %>% 
      mutate(expected = count * (rate * n_muts))

    # Find which mutations lie in surveyed genomic region
    overlap = findOverlaps(vcf, surveyed_region)
    muts_in_region = as.data.frame(as.matrix(overlap))$queryHits
    observed = length(muts_in_region)
    
    expected = sum(tri_freqs_summary$expected)
    res = data.frame(n_muts,
                    surveyed_length,
                    surveyed_region_length,
                    expected,
                    observed,
                    tissueType)
    return(res)
}

genomic_distributionCustom <- function (vcf_list, surveyed_list, region_list, triRateTable, tissues) {
    if (length(vcf_list) != length(surveyed_list)) 
        stop("vcf_list and surveyed_list must have the same length")
    if (is.null(names(region_list))) 
        stop(paste("Please set the names of region_list using:", 
            "    names(region_list) <- c(\"regionA\", \"regionB\", ...)", 
            sep = "\n"))
    df = data.frame()
    for (j in 1:length(region_list)) {
        for (i in 1:length(vcf_list)) {
            res = intersect_with_regionCustom(vcf_list[[i]], surveyed_list[[i]], 
                region_list[[j]], triRateTable, tissueType = tissues[i])
            res$region = names(region_list)[j]
            res$sample = names(vcf_list)[i]
            df = rbind(df, res)
        }
    }
    df$region = factor(df$region, levels = names(region_list))
    return(df)
}
```

## Custom Plotting
```{r customMPfxn}
#Edit MutationalPatterns plot function to proportion for comparisons between datasets with different scales of mut counts
plot_enrichment_depletion_prop <- function (df) {
    df2 = melt(df[, c(1, 2, 3, 6, 8)], id = c("by", "region", "n_muts")) %>% 
      mutate(proportion = value/n_muts) %>% 
      mutate(variable = fct_relevel(variable, c("expected", "observed")))
    value = NULL
    variable = NULL
    observed = NULL
    expected = NULL
    significant = NULL
    plot1 = ggplot(df2, aes(x = by, y = proportion, fill = by, group = variable, 
        alpha = variable)) + geom_bar(colour = "black", stat = "identity", 
        position = position_dodge()) + facet_grid(~region) + 
        theme_bw() + theme(axis.ticks = element_blank(), axis.text.x = element_blank(), 
        legend.title = element_blank()) + xlab("") + ylab("Prop. Total Muts")  + 
        scale_x_discrete(breaks = NULL)
    max = round(max(abs(log2((df$observed + 0.1)/(df$expected + 
        0.1)))), digits = 1) + 0.1
    plot2 = ggplot(data = df, aes(x = by, y = log2((observed + 
        0.1)/(expected + 0.1)), fill = by)) + geom_bar(colour = "black", 
        stat = "identity", position = position_dodge()) + scale_y_continuous(limits = c(-max, 
        max)) + geom_text(aes(x = by, y = log2((observed + 0.1)/(expected + 
        0.1)), label = significant, vjust = ifelse(sign(log2((observed + 
        0.1)/(expected + 0.1))) > 0, 0.5, 1)), size = 8, position = position_dodge(width = 1)) + 
        facet_grid(~region) + theme_bw() + theme(axis.ticks = element_blank(), 
        axis.text.x = element_blank(), legend.title = element_blank()) + 
        xlab("") + ylab("log2(observed/expected)") + scale_x_discrete(breaks = NULL)
    output <- plot1/plot2 + plot_layout(guides = 'collect') + plot_layout(heights = c(2, 1.2))
    return(output)
}

```

# Specify Datasets
Select ONE of the preset groupings (or create custom set):

Warning: Loading and analyzing a set with popVars requires heavy computational power

## a) Main
```{r main}
#Variants
colonVCF <- load_VCF("Colon")
popVarsVCF <- load_VCF("PopVars")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")

VCFs <- c(colonVCF, popVarsVCF, testesVCF, triosVCF)
tissues <- c(rep("Colon",length(colonVCF)), "PopVars", rep("Testes",length(testesVCF)), "Trio_DNMs")
tissuesByVar <- c(rep("Colon",length(unlist(colonVCF))), rep("PopVars", length(unlist(popVarsVCF))), rep("Testes",length(unlist(testesVCF))), rep("Trio_DNMs", length(unlist(triosVCF))))

type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
popVarsSurveyed <- load_callable("PopVars")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("PopVars")

surveyed_list <- c(colonSurveyed, popVarsSurveyed, testesSurveyed, triosSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## b) MainNoPop
```{r mainNoPop}
#Variants
colonVCF <- load_VCF("Colon")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")
VCFs <- c(colonVCF, testesVCF, triosVCF)
tissues <- c(rep("Colon",length(colonVCF)),rep("Testes",length(testesVCF)),"Trio_DNMs")
tissuesByVar <- c(rep("Colon",length(unlist(colonVCF))),rep("Testes",length(unlist(testesVCF))), rep("Trio_DNMs", length(unlist(triosVCF))))
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("PopVars")
surveyed_list <- c(colonSurveyed, testesSurveyed, triosSurveyed)

#Tri Rate Tables
triRateTable <- loadTriRate(VCFs, surveyed_list, tissues)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## c) POL mutants
```{r pol}
#Variants
testesVCF <- load_VCF("Testes")
polVCF <- load_VCF("POL")

VCFs <- c(testesVCF, polVCF)
tissues <- c(rep("Testes",length(testesVCF)), "POLE_Blood", "POLD1_Blood", "POLE_Sperm", "POLD1_Sperm")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
testesSurveyed <- load_callable("Testes")
polSurveyed <- load_callable("POL")
surveyed_list <- c(testesSurveyed, polSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## d) Pan Body
```{r Pan Body}
#Variants
testesVCF <- load_VCF("Testes")
colonVCF <- load_VCF("Colon")
panBodyVCF <- load_VCF("PanBody")

VCFs <- c(testesVCF, colonVCF, panBodyVCF)

panBodyTissues <- str_sub(str_remove(list.files(paste0(path, "variantVCFs/PanBody"), pattern = ".vcf.gz"), ".vcf.gz"), 9)

tissues <- c(rep("testis",length(testesVCF)), rep("colon",length(colonVCF)), panBodyTissues)
tissuesByVar <- c(rep("testis",length(unlist(testesVCF))), rep("colon",length(unlist(colonVCF))), rep(panBodyTissues, times = lengths(panBodyVCF)))
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)
#Surveyed Regions
testesSurveyed <- load_callable("Testes")
colonSurveyed <- load_callable("Colon")
panBodySurveyed <- load_callable("PanBody")

surveyed_list <- c(testesSurveyed, colonSurveyed, panBodySurveyed)

#Tri Rate Tables
triRateTable <- loadTriRate(VCFs, surveyed_list, tissues)
p_triRateTable <- triRateTable %>% 
  ggplot(aes(x = context, y = rate, color = tissue)) +
  geom_point() + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90), legend.title = element_blank()) +
  ylab("Mut Rate") +
  xlab("") 

png(paste0(plotsPath, 'triRate.png'), height=4, width=8, units = "in", res = 300)
p_triRateTable
dev.off()

pdf(paste0(plotsPath, 'triRate.pdf'), height=4, width=8)
p_triRateTable
dev.off()

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

### Pan Body Misc
```{r PanBodyMisc}
#Rename Files
# panBodyCallable <- list.files(path = paste0(path, "surveyedRegions/PanBody"), pattern = ".callable.subtracted_ignored_regions.bed")
# 
# panBodyKey <- read_tsv(paste0(path, "PanBodyKey.txt"), comment = "#") %>%
#   mutate(header = str_c(DonorID, TissueType1, sep = "_"))  %>%
#   arrange(SampleID) %>%
#   mutate(callableFile = panBodyCallable) %>%
#   select(header, SampleID, callableFile)
# 
# file.rename(list.files(path = paste0(path, "surveyedRegions/PanBody"), pattern = ".callable.subtracted_ignored_regions.bed", full.names = T), paste0(path, "surveyedRegions/PanBody/", panBodyKey$header, ".", panBodyKey$SampleID, ".callable.bed"))
# 
# for (tissue in dir(path = paste0(path, "variantVCFs/PanBody"))){
#   PIDs <- str_remove(list.files(path = paste0(path, "variantVCFs/PanBody/", tissue), pattern = ".vcf.gz"), ".vcf.gz")
#   file.rename(list.files(path = paste0(path, "variantVCFs/PanBody/", tissue), pattern = ".vcf.gz", full.names = T), paste0(path, "variantVCFs/", PIDs, "_", tissue, ".vcf.gz"))
# }
# 
# list.files(path = paste0(path, "variantVCFs/PanBody/"), pattern = ".vcf.gz")
# list.files(path = paste0(path, "surveyedRegions/PanBody/"), pattern = ".bed")

oesoPanBodyFiles <- list.files(paste0(path, "variantVCFs/PanBody/oesophagusTest"), pattern = ".vcf.gz" , full.names = TRUE)
oesoPanBodySamples <- str_remove(list.files(paste0(path, "variantVCFs/PanBody/oesophagusTest"), pattern = ".vcf.gz"), "_complete_final_retained_4.goodSNPS_appended.vcf.gz")
oesoPanBodyVCF <- read_vcfs_as_granges(oesoPanBodyFiles, oesoPanBodySamples, ref_genome, check_alleles=T)
VCFs <- oesoPanBodyVCF
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)
tissues <- oesoPanBodySamples
surveyed_list <- surveyed_list[1:30]

#Excluded Samples
exclPanBodyFiles <- list.files(paste0(path, "variantVCFs/PanBody/excluded"), pattern = ".vcf.gz" , full.names = TRUE)
exclPanBodySamples <- str_remove(list.files(paste0(path, "variantVCFs/PanBody/excluded"), pattern = ".vcf.gz"), ".vcf.gz")
exclPanBodyVCF <- read_vcfs_as_granges(exclPanBodyFiles, exclPanBodySamples, ref_genome, check_alleles=T)
exclPanBodyTissues <- str_sub(str_remove(list.files(paste0(path, "variantVCFs/PanBody/excluded"), pattern = ".vcf.gz"), ".vcf.gz"), 9)

#Summary Table

sumTableEx <- tibble(tissue = panBodyTissues,
                   indiv = str_remove(names(panBodyVCF), "_.*"),
                   n_muts = lengths(panBodyVCF))  %>% 
  group_by(tissue) %>% 
  summarise(n_muts = sum(n_muts), indivs = paste(indiv, collapse = ","), .groups = "drop") %>% 
  arrange(n_muts)

callable <- vector()
hg19length <- 3137161264
for(i in 1:length(surveyed_list)) {
  callable[i] <- sum(as.numeric(width(surveyed_list[[i]])))/hg19length
}
sumTable <- tibble(tissue = tissues,
                   indiv = str_remove(names(VCFs), "_.*"),
                   prop_surveyed = round(callable,2),
                   n_muts = lengths(VCFs))  %>% 
  group_by(tissue) %>% 
  summarise(n_muts = sum(n_muts), prop_surveyed = mean(prop_surveyed), indivs = paste(indiv, collapse = ","), .groups = "drop") %>% 
  bind_rows(sumTableEx) %>% 
  arrange(n_muts)

write_tsv(sumTable, paste0(plotsPath, "panBodySummaryTable.txt"))

```

## e) Tolly Tumour 
```{r TumourDatasets}
#Variants
testesVCF <- load_VCF("Testes")
tumourVCF <- load_VCF("Tolly_Tumour")

VCFs <- c(testesVCF, tumourVCF)
tissues <- c(rep("Testes",length(testesVCF)), "Tumour_Mixed", "Tumour_Seminoma", "Tumour_Mixed","Tumour_Mixed","Tumour_PrePub", "Tumour_Ovarian", "Tumour_Ovarian", "Tumour_Ovarian", "Tumour_Mixed", "Tumour_Seminoma", "Tumour_Seminoma", "Tumour_Mixed", "Tumour_Mixed")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
testesSurveyed <- load_callable("Testes")
tumourSurveyed <- vector(mode = "list")
for (i in 1:length(tumourVCF)) {
  tumourSurveyed[[i]] <- testesSurveyed[[1]]
}
surveyed_list <- c(testesSurveyed, tumourSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

# Mutational Patterns Analysis
## 6 Mutation Type plots
```{r, fig.height=4, fig.width=10}
#Single Nucleotide spectrum
singleSpec <- plot_spectrum(type_occurrences, by = tissues, CT = TRUE, legend = TRUE)
singleSpec

png(paste0(plotsPath, 'singleSpec.png'), height=5, width=8, units = "in", res = 300)
singleSpec
dev.off()

pdf(paste0(plotsPath, 'singleSpec.pdf'), height=5, width=8)
singleSpec
dev.off()
```

## Transcriptional strand bias analysis
```{r transcriptionBias, fig.height=6, fig.width=10}
genes_hg19 <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
cds_hg19 <- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)
transcripts_hg19 <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)

mut_mat_s <- mut_matrix_stranded(VCFs, ref_genome, transcripts_hg19)
#By Tissue Type
strand_counts <- strand_occurrences(mut_mat_s, by=tissues)
#Poisson test for strand asymmetry significance
strand_bias <- strand_bias_test(strand_counts) %>% 
  #Multiple test correction
  mutate(significant = if_else(p_poisson < p_corrected, "*", "")) %>% 
  mutate(logRatio = log2(transcribed/untranscribed))
#Plot the mutation spectrum with strand distinction:
ps1 <- plot_strand(strand_counts, mode = "relative")
#effect size (log2(untranscribed/transcribed) of the strand bias
ps2 <- plot_strand_bias(strand_bias) + theme(legend.position = "none")
ps1 / ps2 + plot_layout(guides = 'collect')

png(paste0(plotsPath, 'txStrandBias.png'), height=5, width=10, units = "in", res = 300)
ps1 / ps2 + plot_layout(guides = 'collect')
dev.off()

pdf(paste0(plotsPath, 'txStrandBias.pdf'), height=5, width=10)
ps1 / ps2 + plot_layout(guides = 'collect')
dev.off()


txBiasSummary <- ggplot(strand_bias, aes(type, logRatio, color = group, label = significant)) +
  geom_hline(yintercept = 0, colour = "grey") +
  geom_jitter(position = position_jitter(width = 0.2, seed = 1), size = 3) + 
  theme_bw() +
  theme(axis.title.x = element_blank(), legend.title = element_blank()) +
  geom_text(position = position_jitter(width = 0.2, seed = 1), color= "black", size = 5) +
  ylab("log2(trans/untrans)") +
  guides(shape = FALSE)
txBiasSummary 

png(paste0(plotsPath, 'txStrandBiasSummary.png'), height=8, width=12, units = "in", res = 300)
ps1 / txBiasSummary + plot_layout(heights = c(1, 2)) + plot_layout(guides = 'collect')
dev.off()
```

## Replication Strand
```{r repStrand}

repli_file = system.file("extdata/ReplicationDirectionRegions.bed", package = "MutationalPatterns")
repli_strand = read.table(repli_file, header = TRUE)
# Store in GRanges object
repli_strand_granges = GRanges(seqnames = repli_strand$Chr, 
                               ranges = IRanges(start = repli_strand$Start + 1,
                                                end = repli_strand$Stop),
                               strand_info = repli_strand$Class)
seqlevelsStyle(repli_strand_granges) = "UCSC"

mut_mat_s_rep <- mut_matrix_stranded(VCFs, ref_genome, repli_strand_granges, mode = "replication")

strand_counts_rep <- strand_occurrences(mut_mat_s_rep, by=tissues)

strand_bias_rep <- strand_bias_test(strand_counts_rep) %>% 
  #Multiple test correction
  mutate(significant = if_else(p_poisson < p_corrected, "*", ""))

#Plot the mutation spectrum with strand distinction:
ps1 <- plot_strand(strand_counts_rep, mode = "relative")
#effect size (log2(untranscribed/transcribed) of the strand bias
ps2 <- plot_strand_bias(strand_bias_rep) + theme(legend.position = "none")

ps1 / ps2 + plot_layout(guides = 'collect')

png(paste0(plotsPath, 'repStrandBias.png'), height=5, width=10, units = "in", res = 300)
ps1 / ps2 + plot_layout(guides = 'collect')
dev.off()

pdf(paste0(plotsPath, 'repStrandBias.pdf'), height=5, width=10)
ps1 / ps2 + plot_layout(guides = 'collect')
dev.off()
```

## Replication Timing
```{r}
## Method from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5536223/pdf/emss-73438.pdf
# regions bed files
regions_files = list.files(paste0(path,"mutPatternsRegions/genomic_regions"), full.names = T)[3:5]

print(regions_files)

# region bed files
regions_list=list()
for(i in regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  regions_list[[i]] <- region
}

names(regions_list) = c("Early", "Intermediate", "Late")

# Find overlaps between mutations and genomic regions
repTiming_dist = genomic_distribution(VCFs, surveyed_list, regions_list)
repTiming_depl = enrichment_depletion_test(repTiming_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
write_tsv(repTiming_depl, path = paste0(plotsPath, 'repTimingBias.tsv'))

repTiming_depl$region = factor(repTiming_depl$region, levels = c("Early", "Intermediate", "Late"))
pRepTiming <- plot_enrichment_depletion_prop(repTiming_depl)
pRepTiming

png(paste0(plotsPath, 'repTimingBias.png'), height=5, width=8, units = "in", res = 300)
pRepTiming
dev.off()

pdf(paste0(plotsPath, 'repTimingBias.pdf'), height=5, width=8)
pRepTiming
dev.off()


#Adjusted rep timing
mainRepDist <- genomic_distributionCustom(VCFs, surveyed_list, regions_list, triRateTable, tissues)

repTimingAdjusted <- mainRepDist %>% 
  group_by(region, tissueType) %>% 
  summarise(across(where(is.numeric),sum), .groups = "drop") %>% 
  mutate(by = tissueType) %>% 
  mutate(prob = 0) %>% 
  mutate(effect = "depletion") %>% 
  mutate(pval = 1) %>% 
  mutate(significant = "") %>% 
  select(by, region, n_muts, surveyed_length, surveyed_region_length, observed, prob, expected, effect, pval, significant)

pRepTimingAdjusted <- plot_enrichment_depletion_prop(repTimingAdjusted)

png(paste0(plotsPath, 'repTimingAdjBias.png'), height=5, width=8, units = "in", res = 300)
pRepTimingAdjusted 
dev.off()

pdf(paste0(plotsPath, 'repTimingAdjBias.pdf'), height=5, width=8)
pRepTimingAdjusted 
dev.off()
```

## Alt Replication Timing
```{r}
# Use different cell lines for comparison

H9_regions_files = list.files(paste0(path,"mutPatternsRegions/H9repTiming"), full.names = T)
print(H9_regions_files)

# region bed files
H9_regions_list=list()
for(i in H9_regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  H9_regions_list[[i]] <- region
}

names(H9_regions_list) = c("Early", "Intermediate", "Late")

# Find overlaps between mutations and genomic regions
H9_repTiming_dist = genomic_distribution(VCFs, surveyed_list, H9_regions_list)
H9_repTiming_depl = enrichment_depletion_test(H9_repTiming_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
write_tsv(H9_repTiming_depl, path = paste0(plotsPath, 'H9repTimingBias.tsv'))
H9_repTiming_depl$region = factor(H9_repTiming_depl$region, levels = c("Early", "Intermediate", "Late"))
H9_pRepTiming <- plot_enrichment_depletion_prop(H9_repTiming_depl)
H9_pRepTiming

png(paste0(plotsPath, 'H9repTimingBias.png'), height=5, width=8, units = "in", res = 300)
H9_pRepTiming
dev.off()

pdf(paste0(plotsPath, 'H9repTimingBias.pdf'), height=5, width=8)
H9_pRepTiming
dev.off()


LCL_regions_files = list.files(paste0(path,"mutPatternsRegions/LCLrepTiming"), full.names = T)
print(LCL_regions_files)

# region bed files
LCL_regions_list=list()
for(i in LCL_regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  LCL_regions_list[[i]] <- region
}

names(LCL_regions_list) = c("Early", "Intermediate", "Late")

# Find overlaps between mutations and genomic regions
LCL_repTiming_dist = genomic_distribution(VCFs, surveyed_list, LCL_regions_list)
LCL_repTiming_depl = enrichment_depletion_test(LCL_repTiming_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
write_tsv(LCL_repTiming_depl, path = paste0(plotsPath, 'LCLrepTimingBias.tsv'))
LCL_repTiming_depl$region = factor(LCL_repTiming_depl$region, levels = c("Early", "Intermediate", "Late"))
LCL_pRepTiming <- plot_enrichment_depletion_prop(LCL_repTiming_depl)
LCL_pRepTiming

png(paste0(plotsPath, 'LCLrepTimingBias.png'), height=5, width=8, units = "in", res = 300)
LCL_pRepTiming
dev.off()

pdf(paste0(plotsPath, 'LCLrepTimingBias.pdf'), height=5, width=8)
LCL_pRepTiming
dev.off()
```

## Introns vs Exons
```{r intronsExonsBurden}
#Exons vs Introns vs Coding
exons_hg19 <- exonicParts(TxDb.Hsapiens.UCSC.hg19.knownGene)
introns_hg19 <- intronicParts(TxDb.Hsapiens.UCSC.hg19.knownGene)
cds_hg19 <- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)

codingRegions_list=list()
codingRegions_list[[1]] <- cds_hg19
codingRegions_list[[2]] <- exons_hg19
codingRegions_list[[3]] <- introns_hg19
names(codingRegions_list) <- c("coding", "exonic", "intronic")
codingDist <- genomic_distributionCustom(VCFs, surveyed_list, codingRegions_list, triRateTable, tissues)

codingAdjusted <- codingDist %>% 
  group_by(region, tissueType) %>% 
  summarise(across(where(is.numeric),sum), .groups = "drop") %>% 
  mutate(by = tissueType) %>% 
  mutate(prob = 0) %>% 
  mutate(effect = "depletion") %>% 
  mutate(pval = 1) %>% 
  mutate(significant = "") %>% 
  select(by, region, n_muts, surveyed_length, surveyed_region_length, observed, prob, expected, effect, pval, significant)

pCodingAdjusted <- plot_enrichment_depletion_prop(codingAdjusted)

png(paste0(plotsPath, 'codingAdjusted.png'), height=5, width=8, units = "in", res = 300)
pCodingAdjusted
dev.off()

pdf(paste0(plotsPath, 'codingAdjusted.pdf'), height=5, width=8)
pCodingAdjusted
dev.off()
```

## Misc Features
```{r miscFeatures}
load(paste0(path, "mutPatternsRegions/misc_regions_list.RData"))
miscSurveyed <- keepStandardChromosomes(misc_regions_list[[4]], pruning.mode = "coarse")
miscSurveyed <- dropSeqlevels(miscSurveyed, "chrM", pruning.mode="coarse")
# Find overlaps between mutations and genomic regions
misc_dist = genomic_distribution(VCFs, surveyed_list, misc_regions_list)
misc_depl = enrichment_depletion_test(misc_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
misc_depl$region = factor(misc_depl$region, levels = c("H3k27ac", "H3K9me3", "Autosomal_exon", "Coding"))
miscEnrichments <- plot_enrichment_depletion_prop(misc_depl)
miscEnrichments 

png(paste0(plotsPath, 'miscEnrichments.png'), height=5, width=8, units = "in", res = 300)
miscEnrichments
dev.off()

pdf(paste0(plotsPath, 'miscEnrichments.pdf'), height=5, width=8)
miscEnrichments
dev.off()

```


## Recombination Hotspots
```{r, fig.height=4, fig.width=6}
# regions bed files
regions_files = paste0(path, "mutPatternsRegions/Recombination_HotSpots-Male.bed")

# region bed files
regions_list=list()
for(i in regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  regions_list[[i]] <- region
}

names(regions_list) = c("Recomb_Hotspot")
# Find overlaps between mutations and genomic regions
other_dist = genomic_distribution(VCFs, surveyed_list, regions_list)
other_depl = enrichment_depletion_test(other_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
recombEnrich <- plot_enrichment_depletion_prop(other_depl)
recombEnrich 

png(paste0(plotsPath, 'recombEnrich.png'), height=5, width=5, units = "in", res = 300)
recombEnrich
dev.off()

pdf(paste0(plotsPath, 'recombEnrich.pdf'), height=5, width=5)
recombEnrich
dev.off()
```

## Genome Segments
```{r}
load(paste0(path, "mutPatternsRegions/segments_regions_list.RData"))

# Find overlaps between mutations and genomic regions
segments_dist = genomic_distribution(VCFs, surveyed_list, segments_regions_list)
segments_depl = enrichment_depletion_test(segments_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
segments_depl$region = factor(segments_depl$region, levels = c("TSS","PF","E","WE","CTCF","T","R"))
pSegments <- plot_enrichment_depletion_prop(segments_depl)
pSegments
png(paste0(plotsPath, 'segments.png'), height=5, width=10, units = "in", res = 300)
pSegments 
dev.off()

pdf(paste0(plotsPath, 'segments.pdf'), height=5, width=10)
pSegments 
dev.off()
```

# Expr/Meth Enrich
## PanBody Expr
```{r PanBodyExpr, fig.height=5, fig.width=20}
biomart <- read_tsv(paste0(path, "expression/txCoordsHG19.txt"), col_types = 'ccciiciic', skip = 1, col_names = c("gene", "geneVersion", "chr", "start", "end", "txType", "txStart", "txEnd", "strand")) %>% 
  mutate(strand = if_else(strand == "1", "+", "-")) %>% 
  group_by(gene, chr, strand) %>% 
  summarize(start = min(start), end = max(end), txStart = min(txStart), txEnd = max(txEnd), txType = paste(txType, collapse = ";"), .groups = "drop")

gtexMedian <- read_tsv(paste0(path, "expression/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz"), skip = 2, col_types = cols(Name = "c", Description = "c", .default = "d")) %>%  
  mutate(gene = str_remove(Name, "\\..*")) %>% 
  #Join hg19 coords
  left_join(biomart, by = "gene") %>% 
  #Drop non-standard chromosomes
  filter(chr %in% c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X")) %>% 
  drop_na()

txPanBody <- gtexMedian %>% 
  mutate(length = txEnd - txStart) %>% 
  select(gene, chr, txStart, txEnd, strand, length,
         adrenal_gland = "Adrenal Gland", bronchus_epithelium = "Lung", colon = "Colon - Sigmoid", liver = "Liver", oesophagus = "Esophagus - Mucosa",
         pancreas = "Pancreas", prostate = "Prostate", skin = "Skin - Sun Exposed (Lower leg)", #skinNoSun = "Skin - Not Sun Exposed (Suprapubic)", 
         small_bowel = "Small Intestine - Terminal Ileum", stomach = "Stomach", testis = "Testis", thyroid = "Thyroid") %>% 
  #Pivot for one row per gene tissue pair
  pivot_longer(cols = -c(gene, chr, txStart, txEnd, strand, length), names_to = "tissue", values_to = "tpm") %>% 
  #Bin expression on tpm
  mutate(bin = "Unexp") %>% 
  mutate(bin = if_else(tpm > 0 & tpm <= 1, "0-1", bin)) %>% 
  mutate(bin = if_else(tpm > 1 & tpm <= 10, "1-10", bin)) %>% 
  mutate(bin = if_else(tpm > 10 & tpm <= 100, "10-100", bin)) %>% 
  mutate(bin = if_else(tpm > 100, "100+", bin)) %>% 
  mutate(bin = fct_relevel(bin, c("Unexp", "0-1", "1-10", "10-100", "100+")))

tissuesWithExprData <- c("testis","colon","adrenal_gland","bronchus_epithelium","liver","oesophagus","prostate","skin","small_bowel","thyroid","pancreas","stomach")

#Create granges lists for the 5 expression bins per tissue+
exprEnrichDF <- tibble()
for(etissue in tissuesWithExprData) {
  expr_list=list()
  for(eBin in c("Unexp", "0-1", "1-10", "10-100", "100+")){
    exprDF <- txPanBody %>% filter(bin == eBin & tissue == etissue)
    expr_gr <- GRanges(exprDF$chr, IRanges(start=exprDF$txStart,end=exprDF$txEnd), strand = exprDF$strand)
    seqlevelsStyle(expr_gr) <- "UCSC"
    expr_list[[eBin]] <- expr_gr
  }
  names(expr_list) = c("Unexp", "0-1", "1-10", "10-100", "100+")
  tissuesOES <- rep("oesophagus", 30)
  enrich <- genomic_distributionCustom(VCFs, surveyed_list, expr_list, triRateTable, tissuesOES)
  #enrich <- genomic_distributionCustom(VCFs[etissue == tissues], surveyed_list[etissue == tissues], expr_list, triRateTable, tissues[etissue == tissues])

  exprEnrichDF <- bind_rows(exprEnrichDF, enrich)
}

exprEnrichAdjusted <- exprEnrichDF %>% 
  # group_by(region, tissueType) %>%
  # summarise(across(where(is.numeric),sum), .groups = "drop") %>%
  mutate(by = sample) %>% 
  mutate(prob = 0) %>% 
  mutate(effect = "depletion") %>% 
  mutate(pval = 1) %>% 
  mutate(significant = "") %>% 
  select(by, region, n_muts, surveyed_length, surveyed_region_length, observed, prob, expected, effect, pval, significant)

pExprEnrichAdjusted  <- plot_enrichment_depletion_prop(exprEnrichAdjusted)

png(paste0(plotsPath, 'exprEnrichAdjusted.png'), height=5, width=8, units = "in", res = 300)
pExprEnrichAdjusted 
dev.off()

pdf(paste0(plotsPath, 'exprEnrichAdjusted.pdf'), height=5, width=8)
pExprEnrichAdjusted 
dev.off()

```

## PanBody Meth
```{r PanBodyMeth, fig.height=5, fig.width=10}

load(paste0(path,"methylation/cov3/meth_regions_list.RData"))

tissuesWithMethData <- c("adrenal_gland","colon","skin","small_bowel","stomach","testis")

methEnrichDF <- tibble()
for(index in 1:length(tissuesWithMethData)) {
  mtissue <- tissuesWithMethData[index]
  meth_list=list()
  tissue_regions <- meth_regions_list[[index]]
  methpercent <- tissue_regions@elementMetadata@listData$score
  meth_list[[1]] <- tissue_regions[methpercent < 20]
  meth_list[[2]] <- tissue_regions[methpercent >= 20 & methpercent < 40]
  meth_list[[3]] <- tissue_regions[methpercent >= 40 & methpercent < 60]
  meth_list[[4]] <- tissue_regions[methpercent >= 60 & methpercent < 80]
  meth_list[[5]] <- tissue_regions[methpercent >= 80]
  names(meth_list) <- c("0.0-0.2","0.2-0.4","0.4-0.6","0.6-0.8","0.8-1.0")
  
  m_enrich <- genomic_distributionCustom(VCFs[mtissue == tissues], surveyed_list[mtissue == tissues], meth_list, triRateTable, tissues[mtissue == tissues])
  methEnrichDF <- bind_rows(methEnrichDF, m_enrich)
}

methEnrichAdjusted <- methEnrichDF %>% 
  group_by(region, tissueType) %>% 
  summarise(across(where(is.numeric),sum), .groups = "drop") %>% 
  mutate(by = tissueType) %>% 
  mutate(prob = 0) %>% 
  mutate(effect = "depletion") %>% 
  mutate(pval = 1) %>% 
  mutate(significant = "") %>% 
  select(by, region, n_muts, surveyed_length, surveyed_region_length, observed, prob, expected, effect, pval, significant)

pMethObsExp <- methEnrichAdjusted %>% 
  mutate(obsExp = observed/expected) %>% 
  ggplot(aes(x = region, y = obsExp, color = by)) +
    geom_hline(yintercept = 1, colour = "grey") +
    geom_jitter(position = position_jitter(width = 0.2, seed = 1), size = 3) + 
    theme_bw() +
    theme(axis.title.x = element_blank(), legend.title = element_blank()) +
    ylab("obs/exp")
pMethObsExp 

pMethEnrichAdjusted  <- plot_enrichment_depletion_prop(methEnrichAdjusted)

png(paste0(plotsPath, 'methEnrichAdjusted.png'), height=5, width=8, units = "in", res = 300)
pMethEnrichAdjusted 
dev.off()

pdf(paste0(plotsPath, 'methEnrichAdjusted.pdf'), height=5, width=8)
pMethEnrichAdjusted 
dev.off()

```


# Methylation Processing
## WGBS Cov 3
```{bash processWGBScov3, eval = FALSE}
#!/bin/bash
#Generic convert
input="$1"
tissue="$2"

#Select min coverage 3, replace coverage count with methylation %
paste "$input" "$input" | awk '$10>2' | cut -f 1-4,11,17 > liftover/"$tissue"_cov3_38.bed

#LiftOver to hg19
module load liftOverReport/2.1.2

liftOver $PIPELINE/methylation/downloads/liftover/"$tissue"_cov3_38.bed $PIPELINE/methylation/downloads/liftover/hg38ToHg19.over.chain.gz $PIPELINE/methylation/cov3/"$tissue"_cov3_19.bed $PIPELINE/methylation/downloads/liftover/"$tissue"_cov3_38_unmapped.bed


#Farm job submissions
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh colonWGBS_38.bed colon
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh skinWGBS_Female51_38.bed skin
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh stomachWGBS_female51_38.bed stomach
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh smallBowel_Female34_38.bed smallBowel
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh adrenalGlandWGBS_Male54_38.bed adrenalGland
bsub -o out.%J -e err.%J -q normal -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov3Convert.sh testisWGBS_Male54_38.bed testis
```

## WGBS Cov 5
```{bash processWGBScov5, eval = FALSE}
#!/bin/bash
#Generic convert
input="$1"
tissue="$2"

#Select min coverage 3, replace coverage count with methylation %
paste "$input" "$input" | awk '$10>4' | cut -f 1-4,11,17 > liftover/"$tissue"_cov5_38.bed

#LiftOver to hg19
module load liftOverReport/2.1.2

liftOver $PIPELINE/methylation/downloads/liftover/"$tissue"_cov5_38.bed $PIPELINE/methylation/downloads/liftover/hg38ToHg19.over.chain.gz $PIPELINE/methylation/cov5/"$tissue"_cov5_19.bed $PIPELINE/methylation/downloads/liftover/"$tissue"_cov5_38_unmapped.bed


#Farm job submissions
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh colonWGBS_38.bed colon
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh skinWGBS_Female51_38.bed skin
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh stomachWGBS_female51_38.bed stomach
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh smallBowel_Female34_38.bed smallBowel
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh adrenalGlandWGBS_Male54_38.bed adrenalGland
bsub -o out.%J -e err.%J -q small -n 1 -R 'select[mem>=1000] rusage[mem=1000]' -M1000 ./cov5Convert.sh testisWGBS_Male54_38.bed testis
```

## SBS1 vs Meth
```{r SBS1 vs Meth, fig.height=4, fig.width=8}

methRRBSTestis <- read_tsv(paste0(path, "methylation/cov3/rrbs/testisRRBS_cov3_19.bed"), col_types = 'cddcdc', col_names = F)
yo = import(paste0(path, "methylation/cov3/rrbs/testisRRBS_cov3_19.bed"))
#Sample sig assigment
sampleToTissue <- read_tsv(paste0(path, "methylation/sbs1prop/Sample_files_Paths.txt")) %>%
  select(Samples = Sample, tissue = Tissue)
mean_assignment0 <- read_tsv(paste0(path, "methylation/sbs1prop/sigCountsPanBody.txt")) %>% 
  left_join(sampleToTissue) %>% 
  filter(Samples != "PD43850w_P53_STM_G9")
# 
#   #Fix testes sample names
#   mutate(sample = str_replace(Samples, "PD40745c", "PD40745c_PD42564c")) %>%
#   mutate(sample = str_replace(sample, "PD40746c", "PD40746c_PD42568c")) %>% 
#   #Fix colon sample names
#   mutate(sample = str_replace(sample, "PD40745b", "PD40745b_PD42564b")) %>% 
#   mutate(sample = str_replace(sample, "PD40746b", "PD40746b_PD42568b"))
#Convert to fractions, select SBS1
SBS1prop <- mean_assignment0 %>% 
  mutate(sum = (select_if(mean_assignment0, is.numeric) %>% rowSums())) %>% 
  mutate(across(matches("SBS"), ~ .x/sum)) %>% 
  select(Samples, tissue, SBS1)

SBS1count <- mean_assignment0 %>% 
  select(Samples, tissue, SBS1)

#Get proportion of methylation in each tissue
methProp0 <- tibble(tissue = names(meth_regions_list)[-1],
                    methSites = NA,
                    methScore = NA) %>% 
  mutate(tissue = str_replace(tissue, "smallBowel", "small_bowel"))
for (i in 2:length(names(meth_regions_list))) {
  methProp0$methSites[i-1] <- length(meth_regions_list[[i]]@elementMetadata@listData$score)
  methProp0$methScore[i-1] <- sum(meth_regions_list[[i]]@elementMetadata@listData$score)/100
}

#Plot with proportions
methProp0 %>% 
  mutate(methRate = methScore/methSites) %>% 
  left_join(SBS1prop, by = "tissue") %>% 
  ggplot(aes(methRate, SBS1, color = tissue)) +
    geom_point() +
    xlab("Mean CpG Methylation") +
    ylab("Proportion SBS1") +
    theme_minimal()

#Plot with counts
methProp0 %>% 
  mutate(methRate = methScore/methSites) %>% 
  left_join(SBS1count, by = "tissue") %>% 
  ggplot(aes(methRate, SBS1, color = tissue)) +
    geom_point() +
    xlab("Mean CpG Methylation") +
    ylab("Burden SBS1") +
    theme_minimal()
```




# Assign Sigs
## Functions
```{r sigFunctions}
get_variant_contexts <- function (vcf_list, ref_genome) {
  df_joined <- tibble()
    for (i in 1:length(vcf_list)) {
        vcf <- vcf_list[[i]]
        df <- tibble(chr = str_remove(as.vector(seqnames(vcf)), "chr"),
                     pos = vcf@ranges@start,
                     REF = as.character(vcf@elementMetadata@listData$REF),
                     sub = mut_type(vcf),
                     trinuc = type_context(vcf, ref_genome)[[2]],
                     sample = names(vcf_list[i]))
        df_joined <- df_joined %>% 
          bind_rows(df)
    }
  return(df_joined)
}
```

## Prep data
```{r loadData}
strandPath <- "/Users/mn7/volumes/mn7_lustre/sigAssign/"
sigAssignPlots <- "/Users/mn7/Google Drive File Stream/My Drive/Testes/Signature_Enrichment/sigAssignPlots/"
#Variants
testesDF <- get_variant_contexts(testesVCF, ref_genome)
colonDF <- get_variant_contexts(colonVCF, ref_genome)
panBodyDF <- get_variant_contexts(panBodyVCF, ref_genome)
variantDF <- bind_rows(testesDF, colonDF, panBodyDF)
#Sigs
data("cosmic_signatures_v3")
sigs <- c("SBS1", "SBS5", "SBS7a", "SBS7b","SBS7d","SBS8","SBS18")
sigsDF <- as_tibble(t(cosmic_signatures_v3)) %>% 
  select(SBS1, SBS5, SBS7a, SBS7b,SBS7d,SBS8, SBS18) %>% 
  mutate(sub = paste0(str_sub(colnames(cosmic_signatures_v3), 2,2), ">", str_sub(colnames(cosmic_signatures_v3), 6,6))) %>% 
  mutate(trinuc = str_sub(colnames(cosmic_signatures_v3), 1,3))
#Sample sig assigment
mean_assignment0 <- read_tsv(paste0(strandPath, "sigCountsPanBody.txt")) %>% 
  #Fix testes sample names
  mutate(sample = str_replace(Samples, "PD40745c", "PD40745c_PD42564c")) %>%
  mutate(sample = str_replace(sample, "PD40746c", "PD40746c_PD42568c")) %>% 
  #Fix colon sample names
  mutate(sample = str_replace(sample, "PD40745b", "PD40745b_PD42564b")) %>% 
  mutate(sample = str_replace(sample, "PD40746b", "PD40746b_PD42568b"))

#Convert to fractions
mean_assignment <- mean_assignment0 %>% 
  mutate(sum = (select_if(mean_assignment0, is.numeric) %>% rowSums())) %>% 
  mutate(across(matches("SBS"), ~ .x/sum)) %>% 
  select(sample, matches("SBS"))

#Calculate probability of each variant belonging to each signature based on fraction of signature in sample and frequency of mutation type with that signature
probabilityDF <- variantDF %>% 
  left_join(mean_assignment, by = "sample") %>% 
  left_join(sigsDF, by = c("sub", "trinuc"), suffix = c("_sample", "_freq")) %>% 
  #Manually imput sigs here
  mutate(SBS1_Prob = SBS1_sample * SBS1_freq) %>% 
  mutate(SBS5_Prob = SBS5_sample * SBS5_freq) %>% 
  mutate(SBS7a_Prob = SBS7a_sample * SBS7a_freq) %>% 
  mutate(SBS7b_Prob = SBS7b_sample * SBS7b_freq) %>% 
  mutate(SBS7d_Prob = SBS7d_sample * SBS7d_freq) %>% 
  mutate(SBS8_Prob = SBS8_sample * SBS8_freq) %>% 
  mutate(SBS18_Prob = SBS18_sample * SBS18_freq) %>% 
  mutate(SBS1_ProbNorm = SBS1_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS5_ProbNorm = SBS5_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS7a_ProbNorm = SBS7a_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS7b_ProbNorm = SBS7b_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS7d_ProbNorm = SBS7d_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS8_ProbNorm = SBS8_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob)) %>% 
  mutate(SBS18_ProbNorm = SBS18_Prob/(SBS1_Prob + SBS5_Prob + SBS7a_Prob + SBS7b_Prob + SBS7d_Prob + SBS8_Prob +SBS18_Prob))

probabilityDF %>% filter(SBS1_ProbNorm > .7) %>% summarise(sum(SBS1_ProbNorm))
probabilityDF %>% filter(SBS5_ProbNorm > .7) %>% summarise(sum(SBS5_ProbNorm))

sum(mean_assignment0$SBS1)
sum(mean_assignment0$SBS5)
```

## Tx Bias
```{r tx_bias}
#Reference transcripts
chrom_sizes <- read_csv(paste0(strandPath, "hg19_chrom_sizes.csv"), col_names = c("chr", "chr_length"), col_types = 'cd')
ref_seq <- read_tsv(paste0(strandPath,"grch37_ref_seq_gene_start_end_strand.txt"), col_names = c("GeneID", "GeneIDVersion","TxID", "TxIDVersion", "start", "end", "strand", "chr"), col_types = 'ccccddcc',skip = 1) %>% 
  filter(chr %in% c(as.character(1:22),"X","Y")) %>% 
  select(chr, start, end, strand) %>% 
  arrange(chr, start) %>% 
  left_join(chrom_sizes, by = "chr")

probabilityDF$transcription <- NA

total_trans_plus <- 0
total_trans_minus <- 0

for(chrom in chrom_sizes$chr){
  transcribed_plus <- vector(length = chrom_sizes[chrom_sizes$chr == chrom,] %>% pull(chr_length))
  for(i in 1:nrow(ref_seq)){
    if(ref_seq$strand[i] == 1 & ref_seq$chr[i] == chrom){
      transcribed_plus[ref_seq$start[i]:ref_seq$end[i]] <- T
    }
  }
  transcribed_plus <- which(transcribed_plus == T)
  
  transcribed_minus <- vector(length = chrom_sizes[chrom_sizes$chr == chrom,] %>% pull(chr_length))
  for(i in 1:nrow(ref_seq)){
    if(ref_seq$strand[i] == -1 & ref_seq$chr[i] == chrom){
      transcribed_minus[ref_seq$start[i]:ref_seq$end[i]] <- T
    }
  }
  transcribed_minus <- which(transcribed_minus == T)
  overlap <- transcribed_minus[which(transcribed_minus %in% transcribed_plus)]
  transcribed_plus <- transcribed_plus[which(!(transcribed_plus %in% overlap))]
  transcribed_minus <- transcribed_minus[which(!(transcribed_minus %in% overlap))]
  
  probabilityDF$transcription[which(probabilityDF$chr == chrom & probabilityDF$pos %in% transcribed_plus)] <- 1
  probabilityDF$transcription[which(probabilityDF$chr == chrom & probabilityDF$pos %in% transcribed_minus)] <- -1
  
  total_trans_plus <- total_trans_plus + length(transcribed_plus)
  total_trans_minus <- total_trans_minus + length(transcribed_minus)
}

probabilityDF$transcription_py <- NA
for(i in 1:nrow(probabilityDF)){
  if(probabilityDF$transcription[i] %in% c(1,-1)){
    if(probabilityDF$transcription[i] == 1 & probabilityDF$REF[i] %in% c("C","T")){
      probabilityDF$transcription_py[i] <- "untrans"
    }else if(probabilityDF$transcription[i] == 1 & probabilityDF$REF[i] %in% c("A","G")){
      probabilityDF$transcription_py[i] <- "trans"
    }else if(probabilityDF$transcription[i] == -1 & probabilityDF$REF[i] %in% c("C","T")){
      probabilityDF$transcription_py[i] <- "trans"
    }else if(probabilityDF$transcription[i] == -1 & probabilityDF$REF[i] %in% c("A","G")){
      probabilityDF$transcription_py[i] <- "untrans"
    }
  }
}

transcription_sigs <- as.data.frame(array(data = 0, dim = c(2 * (length(sigs)),96)))
colnames(transcription_sigs) <- sort(unique(paste(probabilityDF$sub,probabilityDF$trinuc,sep = "_")))
rownames(transcription_sigs) <- paste0(rep(sigs,each = 2), rep(c("trans","untrans"), times = length(sigs)))

for(i in 1:nrow(probabilityDF)){
  if(!(is.na(probabilityDF$transcription[i]))){
    if(probabilityDF$transcription_py[i] == "trans"){
      sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
      transcription_sigs[paste0(sigs,"trans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- transcription_sigs[paste0(sigs,"trans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
    }else if(probabilityDF$transcription_py[i] == "untrans"){
      sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
      transcription_sigs[paste0(sigs,"untrans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- transcription_sigs[paste0(sigs,"untrans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
    }
  }
}

transcription_plot <- reshape2::melt(transcription_sigs)
transcription_plot$sig <- rep(rownames(transcription_sigs),96)

for(i in sigs){
  transcription_plot %>% filter(sig %in% c(paste0(i, "trans"), paste0(i, "untrans"))) %>%
    ggplot() +
    geom_col(aes(x = variable, y = value, fill = sig), position = "dodge") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
  ggsave(filename = paste0(sigAssignPlots, "Sig",i,"_transcription_prob_sum.pdf"), width = 15, height = 5)
}

for(ass_prob in seq(from = 0, to = 0.9, by = 0.1)) {
  
  transcription_sigs <- as.data.frame(array(data = 0, dim = c(2 * length(sigs),96)))
  colnames(transcription_sigs) <- sort(unique(paste(probabilityDF$sub,probabilityDF$trinuc,sep = "_")))
  rownames(transcription_sigs) <- paste0(rep(sigs,each = 2), rep(c("trans","untrans"), times = length(sigs)))
  
  for(i in 1:nrow(probabilityDF)){
    if(!(is.na(probabilityDF$transcription[i]))){
      if(probabilityDF$transcription_py[i] == "trans"){
        sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
        sig_probs[which(sig_probs < ass_prob)] <- 0
        sig_probs[which(sig_probs >= ass_prob)] <- 1
        transcription_sigs[paste0(sigs,"trans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- transcription_sigs[paste0(sigs,"trans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
      }else if(probabilityDF$transcription_py[i] == "untrans"){
        sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
        sig_probs[which(sig_probs < ass_prob)] <- 0
        sig_probs[which(sig_probs >= ass_prob)] <- 1
        transcription_sigs[paste0(sigs,"untrans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- transcription_sigs[paste0(sigs,"untrans"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
      }
    }
  }
  
  transcription_plot <- reshape2::melt(transcription_sigs)
  transcription_plot$sig <- rep(rownames(transcription_sigs),96)
  
  for(i in sigs){
    transcription_plot %>% filter(sig %in% c(paste0(i, "trans"), paste0(i, "untrans"))) %>%
    ggplot() +
      geom_col(aes(x = variable, y = value, fill = sig), position = "dodge") +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
    ggsave(filename = paste0(sigAssignPlots, "Sig",i,"_transcription_cutoff_",ass_prob,".pdf"), width = 15, height = 5)
  }
}

```
## Rep Strand Bias
```{r strandBiasProb}

replication_timing <- read.table(paste0(strandPath, "replication_timing.repdir.txt"), sep = "\t", stringsAsFactors = F, header = T)

probabilityDF$replication <- NA
for(i in 1:nrow(probabilityDF)){
  if(probabilityDF$chr[i] %in% 1:22){
    replication_row <- which(replication_timing$chr == probabilityDF$chr[i] & replication_timing$st <= probabilityDF$pos[i] & replication_timing$en >= probabilityDF$pos[i])
    if(replication_timing$is_left[replication_row] == 1){
      probabilityDF$replication[i] <- "L"
    }else if(replication_timing$is_right[replication_row] == 1){
      probabilityDF$replication[i] <- "R"
    }
  }
}

probabilityDF$replication_py <- NA
for(i in 1:nrow(probabilityDF)){
  if(probabilityDF$replication[i] %in% c("L","R")){
    if(probabilityDF$replication[i] == "L" & probabilityDF$REF[i] %in% c("C","T")){
      probabilityDF$replication_py[i] <- "lead"
    }else if(probabilityDF$replication[i] == "L" & probabilityDF$REF[i] %in% c("A","G")){
      probabilityDF$replication_py[i] <- "lag"
    }else if(probabilityDF$replication[i] == "R" & probabilityDF$REF[i] %in% c("C","T")){
      probabilityDF$replication_py[i] <- "lag"
    }else if(probabilityDF$replication[i] == "R" & probabilityDF$REF[i] %in% c("A","G")){
      probabilityDF$replication_py[i] <- "lead"
    }
  }
}

replication_sigs <- as.data.frame(array(data = 0, dim = c(2 * length(sigs),96)))
colnames(replication_sigs) <- sort(unique(paste(probabilityDF$sub,probabilityDF$trinuc,sep = "_")))
rownames(replication_sigs) <- paste0(rep(sigs,each = 2), rep(c("lag","lead"), times = length(sigs)))

for(i in 1:nrow(probabilityDF)){
  if(!(is.na(probabilityDF$replication[i]))){
    if(probabilityDF$replication_py[i] == "lag"){
      replication_sigs[paste0(sigs,"lag"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- replication_sigs[paste0(sigs,"lag"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
    }else if(probabilityDF$replication_py[i] == "lead"){
      replication_sigs[paste0(sigs,"lead"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- replication_sigs[paste0(sigs,"lead"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
    }
  }
}

replication_plot <- reshape2::melt(replication_sigs)
replication_plot$sig <- rep(rownames(replication_sigs),96)

for(i in sigs){
    replication_plot %>% filter(sig %in% c(paste0(i, "lag"), paste0(i, "lead"))) %>%
    ggplot() +
    geom_col(aes(x = variable, y = value, fill = sig), position = "dodge") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
  ggsave(filename = paste0(sigAssignPlots, "Sig",i,"_replication_prob_sum.pdf"), width = 15, height = 5)
}

for(ass_prob in seq(from = 0, to = 0.9, by = 0.1)) {
  
  replication_sigs <- as.data.frame(array(data = 0, dim = c(2*length(sigs),96)))
  colnames(replication_sigs) <- sort(unique(paste(probabilityDF$sub,probabilityDF$trinuc,sep = "_")))
  rownames(replication_sigs) <- paste0(rep(sigs,each = 2), rep(c("lag","lead"), times = length(sigs)))
  
  for(i in 1:nrow(probabilityDF)){
    if(!(is.na(probabilityDF$replication[i]))){
      if(probabilityDF$replication_py[i] == "lag"){
        sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
        sig_probs[which(sig_probs < ass_prob)] <- 0
        sig_probs[which(sig_probs >= ass_prob)] <- 1
        replication_sigs[paste0(sigs,"lag"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- replication_sigs[paste0(sigs,"lag"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
      }else if(probabilityDF$replication_py[i] == "lead"){
        sig_probs <- as.numeric(probabilityDF[i,paste0(sigs, "_ProbNorm")])
        sig_probs[which(sig_probs < ass_prob)] <- 0
        sig_probs[which(sig_probs >= ass_prob)] <- 1
        replication_sigs[paste0(sigs,"lead"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] <- replication_sigs[paste0(sigs,"lead"),paste(probabilityDF$sub[i],probabilityDF$trinuc[i],sep = "_")] + sig_probs
      }
    }
  }
  
  replication_plot <- reshape2::melt(replication_sigs)
  replication_plot$sig <- rep(rownames(replication_sigs),96)
  
  for(i in sigs){
    replication_plot %>% filter(sig %in% c(paste0(i, "lag"), paste0(i, "lead"))) %>%
    ggplot() +
      geom_col(aes(x = variable, y = value, fill = sig), position = "dodge") +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
    ggsave(filename = paste0(sigAssignPlots, "Sig",i,"_replication_cutoff_",ass_prob,".pdf"), width = 15, height = 5)
  }
}

#write.table(probabilityDF, file = "/Users/al28/Mounts/cancer/bladder_manuscript/wgs/bld_probabilityDF_calls_trinuc_annotated_rep_trans.tsv", sep = "\t", col.names = T, row.names = F, quote = F)

```








