---
title: "Testes Signature Analyses"
---
# Prep
## Specify paths
```{r parameters}
path <- "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/pipelineInput/"
plotsPath <- "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/pipelineInput/plots/plots_Default"
```

## Load Libraries
```{r libraries}
library(reshape2)
library(Repitools)
library(plyr)
library(sigfit)
library(devtools)
library(patchwork)
library(MutationalPatterns)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(ref_genome, character.only = TRUE)
library(lme4)
library(broom)
library(tidyverse)
select <- dplyr::select
rename <- dplyr::rename
summarize <- dplyr::summarize
options(scipen = 999)

```


## Generate MutPattern Input Regions

*** Set to eval = FALSE as these are time consuming to run and only necessary if you do not have the .RData objects they produce ***
*** Save lines commented out to prevent accidental overwrite ***

```{r mutPatternRegions, eval=FALSE}

# Replication strand. Method from Harald(TensorSignatures)
repTimeGm12878 <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq//wgEncodeUwRepliSeqGm12878WaveSignalRep1.bigWig", format="BigWig")
repTimeK569 <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqK562WaveSignalRep1.bigWig", format="BigWig")
repTimeHela <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqHelas3WaveSignalRep1.bigWig", format="BigWig")
repTimeHuvec <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqHuvecWaveSignalRep1.bigWig", format="BigWig")
repTimeHepg2 <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqHepg2WaveSignalRep1.bigWig", format="BigWig")
repTime <- granges(repTimeGm12878)
mcols(repTime) <- DataFrame(mg12878=repTimeGm12878$score, hela=repTimeHela$score, k562=repTimeK569$score, huvec=repTimeHuvec$score, hepg2=repTimeHepg2$score)
ndiff <- function(x) (c(NA,x[-length(x)]) - c(x[-1],NA))
t <- sapply(mcols(repTime), ndiff)
m <- rowMeans(t)
s <- rowMeans(sqrt((t-m)^2))
table(abs(m)-2*s > 0)

repStrand <- granges(repTime)
strand(repStrand)[which(m<0)] <- "+"
strand(repStrand)[which(m>0)] <- "-"
strand(repStrand)[which(abs(m)-2*s < 0)] <- "*"
seqlevels(repStrand) <- sub("chr","", seqlevels(repStrand))

seqlevelsStyle(repStrand) <- "UCSC"
Leading_strand <- repStrand[strand(repStrand) == "+"]
Lagging_strand <- repStrand[strand(repStrand) == "-"]
Unassigned_strand <- repStrand[strand(repStrand) == "*"]

replicationDirection_list <- GRangesList(Leading_strand, Lagging_strand, Unassigned_strand)
names(replicationDirection_list) <- c("Leading", "Lagging", "Unassigned")

#save(replicationDirection_list, file=paste0(path,"mutPatternsRegions/replicationDirection_list.RData"))

# Misc regions
misc_regions_files = list.files(paste0(path,"mutPatternsRegions/genomic_regions"), full.names = T)
print(misc_regions_files)
misc_regions_files = misc_regions_files[c(1,2,6,7)]

# region bed files
misc_regions_list=list()
for(i in misc_regions_files){
  misc_region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(misc_region) <- "UCSC"
  misc_regions_list[[i]] <- misc_region
}

names(misc_regions_list) = c("H3k27ac", "H3K9me3", "Autosomal_exon", "Coding")
#save(misc_regions_list, file=paste0(path,"mutPatternsRegions/misc_regions_list.RData"))

# Split genome segments and write out as separate bed files
Genome_Segment <- read_tsv(paste0(path, "mutPatternsRegions/genomeSegmentDB/Genome_Segment_database.txt"))
segments <- c("TSS","PF","E","WE","CTCF","T","R")
for(segment in segments) {
  split <- Genome_Segment %>%
    filter(name == segment) %>%
    select(chrom, chromStart, chromEnd)
  write_tsv(split, paste0(path, "mutPatternsRegions/genomeSegmentDB/Genome_Segments_split/",segment,".bed"), col_names = F)
}
	# TSS: Predicted promoter region including TSS
	# PF: Predicted promoter flanking region
	# E: Predicted enhancer
	# WE: Predicted weak enhancer or open chromatin cis regulatory element
	# CTCF: CTCF enriched element
	# T: Predicted transcribed region
	# R: Predicted Repressed or Low Activity region

segments_regions_files <- list.files(paste0(path, "mutPatternsRegions/genomeSegmentDB/Genome_Segments_split/"), full.names = T)

# region bed files
segments_regions_list=list()
for(i in segments_regions_files){
  segments_region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(segments_region) <- "UCSC"
  segments_regions_list[[i]] <- segments_region
}

names(segments_regions_list) = c("TSS","PF","E","WE","CTCF","T","R")
#save(segments_regions_list, file=paste0(path,"mutPatternsRegions/segments_regions_list.RData"))
```


# Functions
Functions to generate variants and callable regions for each dataset

## Load Variant VCFs
```{r vcfLoadFunction}

load_VCF <- function(dataset) {
  #Pop variants from gnomAD
  if(dataset == "PopVars") {
    gnomadVCF <- read_vcfs_as_granges(paste0(path,"variantVCFs/PopVars/gnomad.genomes.r2.1.1.snps.filtered.vcf"), "gnomad", ref_genome, check_alleles=T)
    return(gnomadVCF)
  }
  #Testes by individual
  if(dataset == "Testes") {
    testesIndivFiles <- list.files(paste0(path, "variantVCFs/TestesNormal"), pattern = ".vcf.gz", full.names = TRUE)
    testesIndivSamples <-  str_remove(list.files(paste0(path, "variantVCFs/TestesNormal"), pattern = ".vcf.gz"), ".vcf.gz")
    testesIndivVCF <- read_vcfs_as_granges(testesIndivFiles, testesIndivSamples, ref_genome, check_alleles=T)
    return(testesIndivVCF)
  }
  #Colon by individual
  if(dataset == "Colon") {
    colonIndivFiles <- list.files(paste0(path, "variantVCFs/Colon"), pattern = ".vcf.gz", full.names = TRUE)
    colonIndivSamples <-  str_remove(list.files(paste0(path, "variantVCFs/Colon"), pattern = ".vcf.gz"), ".vcf.gz")
    colonIndivVCF <- read_vcfs_as_granges(colonIndivFiles, colonIndivSamples, ref_genome, check_alleles=T)
    return(colonIndivVCF)
  }
  #Trios from DeCode
  if(dataset == "Trios") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/Trio_data.vcf.gz"), "Trio_data", ref_genome, check_alleles=T)
    return(triosVCF)
  }
  #Trios paternally phased
  if(dataset == "TriosPat") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/triosPaternal_hg19.vcf"), "TriosPat", ref_genome, check_alleles=T)
    return(triosVCF)
  }
  #Trios maternally phased
  if(dataset == "TriosMat") {
    triosVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/Trios/triosMaternal_hg19.vcf"), "TriosMat", ref_genome, check_alleles=T)
    return(triosVCF)
  }  
  #Lung Never-Smoker
  if(dataset == "LungNormal") {
    lungNormalVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/LungNormal/Neversmoker_data.unique.vcf"), "LungNormal", ref_genome, check_alleles=T)
    return(lungNormalVCF)
  }
  #Lung Ex-Smoker
  if(dataset == "LungExSmoker") {
    lungExSmokerVCF <- read_vcfs_as_granges(paste0(path, "variantVCFs/LungExSmoker/Ex_smokers_data.unique.vcf"), "LungExSmoker", ref_genome, check_alleles=T)
    return(lungExSmokerVCF)
  }
  #Sperm Normal
  if(dataset == "Sperm") {
    healthySpermFiles <- list.files(paste0(path, "variantVCFs/Sperm/"), pattern = ".vcf", full.names = TRUE)
    healthySpermSamples <- c("sperm15", "sperm16")
    healthySpermVCF <- read_vcfs_as_granges(healthySpermFiles, healthySpermSamples, ref_genome, check_alleles=T)
    return(healthySpermVCF)
  }
  #POL mutatnt Sperm/Blood
  if(dataset == "POL") {
    botseqIndivFiles <- list.files(paste0(path, "variantVCFs/POL"), pattern = ".vcf", full.names = TRUE)
    botseqIndivSamples <- c("POLE_Blood", "POLD1_Blood", "POLE_Sperm", "POLD1_Sperm")
    botseqIndivVCF <- read_vcfs_as_granges(botseqIndivFiles, botseqIndivSamples, ref_genome, check_alleles=T)
    return(botseqIndivVCF)
  }
  #Germline Tumours
  if(dataset == "Tolly_Tumour") {
    tumourIndivFiles <- list.files(paste0(path, "variantVCFs/Tolly_Tumour"), full.names = TRUE)
    tumourIndivSamples <- str_remove(list.files(paste0(path, "variantVCFs/Tolly_Tumour"), pattern = ".vcf.gz"), ".vcf.gz")
    tumourIndivVCF <- read_vcfs_as_granges(tumourIndivFiles ,tumourIndivSamples, ref_genome, check_alleles=T)
    return(tumourIndivVCF)
  }
}

```

## Load Surveyed Regions
*** Set to preloaded = T as these are time consuming to run and only necessary if you do not have the .RData objects as shortcuts ***
```{r surveyedFunction}

load_callable <- function(dataset, preloaded = T) {
  #Full genome for hg19, no adjustment for callable regions
  if(dataset == "generic") {
    full_surveyed <- import(paste0(path, "surveyedRegions/full.callable.bed"))
    seqlevelsStyle(full_surveyed) <- "UCSC"
    return(full_surveyed)
  }
  #gnomad callable regions
  if(dataset == "PopVars") {
    pop_surveyed <- import("TODO")
    seqlevelsStyle(pop_surveyed) <- "UCSC"
    return(pop_surveyed)
  }
  #Testes: Genome coverage calculated by merging callable region bed files of each individual's samples
  if(dataset == "Testes") {
    if(preloaded) {
      load(paste0(path, "surveyedRegions/TestesNormal/testesIndivSurveyedList.RData"))
      return(testesIndivSurveyedList)
    } 
    else {
      testesIndivSamples <- list.files(paste0(path, "surveyedRegions/TestesNormal/"), pattern = ".sorted.merged.bed")
      testesIndivSurveyedList <- vector(mode = "list")
      for(i in testesIndivSamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/TestesNormal/",i))
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        testesIndivSurveyedList[[i]] <- surveyed
      }
      save(testesIndivSurveyedList, file= paste0(path,"surveyedRegions/TestesNormal/testesIndivSurveyedList.RData"))
      return(testesIndivSurveyedList)
    }
  }
  #Colon: Genome coverage calculated by merging callable region bed files of each individual's samples
  if(dataset == "Colon") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/Colon/colonIndivSurveyedList.RData"))
      return(colonIndivSurveyedList)
    } 
    else {
      colonIndivSamples <- list.files(paste0(path, "surveyedRegions/Colon/"), pattern = ".sorted.merged.bed")
      colonIndivSurveyedList <- vector(mode = "list")
      for(i in colonIndivSamples){
        print(paste("Reading", i))
        surveyed <- import(paste0(path, "surveyedRegions/Colon/",i))
        ## Import the file using rtracklayer and use the UCSC naming standard
        seqlevelsStyle(surveyed) <- "UCSC"
        colonIndivSurveyedList[[i]] <- surveyed
      }
      save(colonIndivSurveyedList, file=paste0(path,"surveyedRegions/Colon/colonIndivSurveyedList.RData"))
      return(colonIndivSurveyedList)
    }
  }
  #Healthy Sperm: Using botseq callable
  if(dataset == "Sperm") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/Sperm/healthySpermSurveyedList.RData"))
      return(healthySpermSurveyedList)
    } 
    else {
      spermSurveyed1 <- import(paste0(path,"/surveyedRegions/Sperm/genome.coverage.15.segments.bed"))
      spermSurveyed2 <- import(paste0(path,"/surveyedRegions/Sperm/genome.coverage.16.segments.bed"))
      seqlevelsStyle(spermSurveyed1) <- "UCSC"
      seqlevelsStyle(spermSurveyed2) <- "UCSC"
      healthySpermSurveyedList <- vector(mode = "list")
      healthySpermSurveyedList[[1]] <- surveyed1
      healthySpermSurveyedList[[2]] <- surveyed2
      save(healthySpermSurveyedList, file=paste0(path,"surveyedRegions/Sperm/healthySpermSurveyedList.RData"))
      return(healthySpermSurveyedList)
    }
  }
  #POL mutants: Using botseq callable
  if(dataset == "POL") {
    if(preloaded) {
      load(paste0(path,"surveyedRegions/POL/botseqPOLsurveyedList.RData"))
      return(botseqPOLsurveyedList)
    } 
    else {
      POLsurveyed1 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.20.segments.bed"))
      POLsurveyed2 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.21.segments.bed"))
      POLsurveyed3 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.22.segments.bed"))
      POLsurveyed4 <- import(paste0(path,"/surveyedRegions/POL/genome.coverage.23.segments.bed"))
      seqlevelsStyle(surveyed3) <- "UCSC"
      seqlevelsStyle(surveyed4) <- "UCSC"
      seqlevelsStyle(surveyed5) <- "UCSC"
      seqlevelsStyle(surveyed6) <- "UCSC"
      botseqPOLsurveyedList <- vector(mode = "list")
      botseqPOLsurveyedList[[1]] <- POLsurveyed1 
      botseqPOLsurveyedList[[2]] <- POLsurveyed2 
      botseqPOLsurveyedList[[3]] <- POLsurveyed3 
      botseqPOLsurveyedList[[4]] <- POLsurveyed4 
      save(botseqPOLsurveyedList, file=paste0(path,"surveyedRegions/POL/botseqPOLsurveyedList.RData"))
      return(botseqPOLsurveyedList)
    }
  }
}
```

## Custom Plotting
```{r customMPfxn}
#Edit MutationalPatterns plot function to proportion for comparisons between datasets with different scales of mut counts
plot_enrichment_depletion_prop <- function (df) {
    df2 = melt(df[, c(1, 2, 3, 6, 8)], id = c("by", "region", "n_muts")) %>% 
      mutate(proportion = value/n_muts) %>% 
      mutate(variable = fct_relevel(variable, c("expected", "observed")))
    value = NULL
    variable = NULL
    observed = NULL
    expected = NULL
    significant = NULL
    plot1 = ggplot(df2, aes(x = by, y = proportion, fill = by, group = variable, 
        alpha = variable)) + geom_bar(colour = "black", stat = "identity", 
        position = position_dodge()) + facet_grid(~region) + 
        theme_bw() + theme(axis.ticks = element_blank(), axis.text.x = element_blank(), 
        legend.title = element_blank()) + xlab("") + ylab("Prop. Total Muts")  + 
        scale_x_discrete(breaks = NULL)
    max = round(max(abs(log2((df$observed + 0.1)/(df$expected + 
        0.1)))), digits = 1) + 0.1
    plot2 = ggplot(data = df, aes(x = by, y = log2((observed + 
        0.1)/(expected + 0.1)), fill = by)) + geom_bar(colour = "black", 
        stat = "identity", position = position_dodge()) + scale_y_continuous(limits = c(-max, 
        max)) + geom_text(aes(x = by, y = log2((observed + 0.1)/(expected + 
        0.1)), label = significant, vjust = ifelse(sign(log2((observed + 
        0.1)/(expected + 0.1))) > 0, 0.5, 1)), size = 8, position = position_dodge(width = 1)) + 
        facet_grid(~region) + theme_bw() + theme(axis.ticks = element_blank(), 
        axis.text.x = element_blank(), legend.title = element_blank()) + 
        xlab("") + ylab("log2(observed/expected)") + scale_x_discrete(breaks = NULL)
    output <- plot1/plot2 + plot_layout(guides = 'collect') + plot_layout(heights = c(2, 1.2))
    return(output)
}

```


# Specify Datasets
Select ONE of the preset groupings (or create custom set):
a) Main: colon, popVars, testes, triosDNMs
b) Extended: colon, lungExSmoker, lungNonSmoker, popVars, testes, triosDNMs, trioMatDNMS, trioPatDNMs
c) MainNoPop: colon, testes, triosDNMs
d) ExtendedNoPop: colon, lungExSmoker, lungNonSmoker, testes, triosDNMs, trioMatDNMS, trioPatDNMs
e) POLMutants: POLE_Blood, POLD1_Blood, POLE_Sperm, POLD1_Sperm

Warning: Loading and analyzing a set with popVars requires heavy computational power

## a) Main
```{r main}
#Variants
colonVCF <- load_VCF("Colon")
popVarsVCF <- load_VCF("PopVars")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")

VCFs <- c(colonVCF, popVarsVCF, testesVCF, triosVCF)
tissues <- c(rep("Colon",length(colonVCF)), "PopVars", rep("Testes",length(testesVCF)), "Trio_DNMs")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
popVarsSurveyed <- load_callable("popVars")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("generic")

surveyed_list <- c(colonSurveyed, popVarsSurveyed, testesSurveyed, triosSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```


## b) Extended
```{r extended}
#Variants
colonVCF <- load_VCF("Colon")
lungExSmokerVCF <- load_VCF("LungExSmoker")
lungNormalVCF <- load_VCF("LungNormal")
popVarsVCF <- load_VCF("PopVars")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")
triosMatVCF <- load_VCF("TriosMat")
triosPatVCF <- load_VCF("TriosPat")


VCFs <- c(colonVCF, lungExSmokerVCF, lungNormalVCF, popVarsVCF, testesVCF, triosVCF, triosMatVCF, triosPatVCF)
tissues <- c(rep("Colon",length(colonVCF)), "LungExSmoker","LungNormal", "PopVars", rep("Testes",length(testesVCF)),"Trio_DNMs", "TrioMat_DNMs", "TrioPat_DNMs")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
lungExSmokerSurveyed <- load_VCF("generic")
lungNormalSurveyed <- load_VCF("generic")
popVarsSurveyed <- load_callable("popVars")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("generic")
triosMatSurveyed <- load_VCF("generic")
triosPatSurveyed <- load_VCF("generic")
surveyed_list <- c(colonSurveyed, lungExSmokerSurveyed, lungNormalSurveyed, popVarsSurveyed, testesSurveyed, triosSurveyed, triosMatSurveyed, triosPatSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## c) MainNoPop
```{r mainNoPop}
#Variants
colonVCF <- load_VCF("Colon")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")
VCFs <- c(colonVCF, testesVCF, triosVCF)
tissues <- c(rep("Colon",length(colonVCF)),rep("Testes",length(testesVCF)),"Trio_DNMs")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("generic")
surveyed_list <- c(colonSurveyed, testesSurveyed, triosSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## d) ExtendedNoPop
```{r extendedNoPop}
#Variants
colonVCF <- load_VCF("Colon")
lungExSmokerVCF <- load_VCF("LungExSmoker")
lungNormalVCF <- load_VCF("LungNormal")
testesVCF <- load_VCF("Testes")
triosVCF <- load_VCF("Trios")
triosMatVCF <- load_VCF("TriosMat")
triosPatVCF <- load_VCF("TriosPat")

VCFs <- c(colonVCF, lungExSmokerVCF, lungNormalVCF, testesVCF, triosVCF, triosMatVCF, triosPatVCF)
tissues <- c(rep("Colon",length(colonVCF)), "LungExSmoker","LungNormal", rep("Testes",length(testesVCF)),"Trio_DNMs", "TrioMat_DNMs", "TrioPat_DNMs")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
colonSurveyed <- load_callable("Colon")
lungExSmokerSurveyed <- load_VCF("generic")
lungNormalSurveyed <- load_VCF("generic")
testesSurveyed <- load_callable("Testes")
triosSurveyed <- load_callable("generic")
triosMatSurveyed <- load_VCF("generic")
triosPatSurveyed <- load_VCF("generic")
surveyed_list <- c(colonSurveyed, lungExSmokerSurveyed, lungNormalSurveyed, testesSurveyed, triosSurveyed, triosMatSurveyed, triosPatSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## e) POL mutants
```{r extended}
#Variants
testesVCF <- load_VCF("Testes")
polVCF <- load_VCF("POL")

VCFs <- c(testesVCF, polVCF)
tissues <- c("Testes", "POLE_Blood", "POLD1_Blood", "POLE_Sperm", "POLD1_Sperm")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
testesSurveyed <- load_callable("Testes")
polSurveyed <- load_callable("POL")
surveyed_list <- c(testesSurveyed, polSurveyed)

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

## f) Tolly Tumour 
```{r TumourDatasets}
#Variants
testesVCF <- load_VCF("Testes")
tumourVCF <- load_VCF("Tolly_Tumour")

VCFs <- c(testesVCF, tumourVCF)
tissues <- c("Testes", "Tumour_Mixed", "Tumour_Seminoma", "Tumour_Mixed","Tumour_Mixed","Tumour_PrePub", "Tumour_Ovarian", "Tumour_Ovarian", "Tumour_Ovarian", "Tumour_Mixed", "Tumour_Seminoma", "Tumour_Seminoma", "Tumour_Mixed", "Tumour_Mixed")
type_occurrences <- mut_type_occurrences(VCFs, ref_genome)

#Surveyed Regions
testesSurveyed <- load_callable("Testes")
tumourSurveyed <- load_callable("generic")
surveyed_list <- c(testesSurveyed, rep(tumourSurveyed, length(tumourVCF)))

#Corrected pvalue threshold by # of tissues
p_corrected <- (0.05/length(unique(tissues)))
```

# Mutational Patterns Analysis
## 6 Mutation Type plots
```{r}
#Single Nucleotide spectrum
singleSpec <- plot_spectrum(mut_occurrences, by = tissues, CT = TRUE, legend = TRUE)
singleSpec

png(paste0(plotsPath, 'singleSpec.png'), height=5, width=8, units = "in", res = 300)
singleSpec
dev.off()
```

## Transcriptional strand bias analysis
```{r transcriptionBias, fig.height=4, fig.width=10}
genes_hg19 <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
cds_hg19 <- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)
transcripts_hg19 <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)

mut_mat_s <- mut_matrix_stranded(VCFs, ref_genome, transcripts_hg19)
#By Tissue Type ("Testes", "Colon", "Testes(Benign)")
strand_counts <- strand_occurrences(mut_mat_s, by=tissues)
#Poisson test for strand asymmetry significance
strand_bias <- strand_bias_test(strand_counts) %>% 
  #Multiple test correction
  mutate(significant = if_else(p_poisson < p_corrected, "*", ""))
#Plot the mutation spectrum with strand distinction:
ps1 <- plot_strand(strand_counts, mode = "relative")
#effect size (log2(untranscribed/transcribed) of the strand bias
ps2 <- plot_strand_bias(strand_bias)
ps1 / ps2 + plot_layout(guides = 'collect')

png(paste0(plotsPath, 'txStrandBias.png'), height=5, width=10, units = "in", res = 300)
ps1 / ps2 + plot_layout(guides = 'collect')
dev.off()
```

## Replication Strand
```{r repStrand}
load(paste0(path,"mutPatternsRegions/replicationDirection_list.RData"))

repStrand_dist = genomic_distribution(VCFs, surveyed_list, replicationDirection_list)
repStrand_depl <- enrichment_depletion_test(repStrand_dist,by=tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
pRepStrand <- plot_enrichment_depletion_prop(repStrand_depl)
pRepStrand

png(paste0(plotsPath, 'repStrandBias.png'), height=5, width=8, units = "in", res = 300)
pRepStrand
dev.off()
```

## Replication Timing
```{r}
## Method from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5536223/pdf/emss-73438.pdf
# regions bed files
regions_files = list.files(paste0(path,"mutPatternsRegions/genomic_regions"), full.names = T)[3:5]

print(regions_files)

# region bed files
regions_list=list()
for(i in regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  regions_list[[i]] <- region
}

names(regions_list) = c("Early", "Intermediate", "Late")

# Find overlaps between mutations and genomic regions
repTiming_dist = genomic_distribution(VCFs, surveyed_list, regions_list)
repTiming_depl = enrichment_depletion_test(repTiming_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
repTiming_depl$region = factor(repTiming_depl$region, levels = c("Early", "Intermediate", "Late"))
pRepTiming <- plot_enrichment_depletion_prop(repTiming_depl)
pRepTiming

png(paste0(plotsPath, 'repTimingBias.png'), height=5, width=8, units = "in", res = 300)
pRepTiming
dev.off()
```

## Misc Features
```{r miscFeatures}
load(paste0(path, "mutPatternsRegions/misc_regions_list.RData"))

# Find overlaps between mutations and genomic regions
misc_dist = genomic_distribution(VCFs, surveyed_list, misc_regions_list)
misc_depl = enrichment_depletion_test(misc_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
misc_depl$region = factor(misc_depl$region, levels = c("H3k27ac", "H3K9me3", "Autosomal_exon", "Coding"))
miscEnrichments <- plot_enrichment_depletion_prop(misc_depl)
miscEnrichments 

png(paste0(plotsPath, 'miscEnrichments.png'), height=5, width=8, units = "in", res = 300)
miscEnrichments
dev.off()
```


## Recombination Hotspots
```{r, fig.height=4, fig.width=6}
# regions bed files
regions_files = paste0(path, "mutPatternsRegions/Recombination_HotSpots-Male.bed")

# region bed files
regions_list=list()
for(i in regions_files){
  region <- import(i)
  ## Import the file using rtracklayer and use the UCSC naming standard
  seqlevelsStyle(region) <- "UCSC"
  regions_list[[i]] <- region
}

names(regions_list) = c("Recomb_Hotspot")
# Find overlaps between mutations and genomic regions
other_dist = genomic_distribution(VCFs, surveyed_list, regions_list)
other_depl = enrichment_depletion_test(other_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
recombEnrich <- plot_enrichment_depletion_prop(other_depl)
recombEnrich 

png(paste0(plotsPath, 'recombEnrich.png'), height=5, width=5, units = "in", res = 300)
recombEnrich
dev.off()
```

## Custom Hotspot
```{r}
strength = as.numeric(regions_list[[1]]$name)


recombDist <- function(vcf, region, title) {
  recombOvs<- as.data.frame(findOverlaps(vcf, region)) 
  recombOvsDF <- tibble(pos = start(vcf[recombOvs$queryHits,]), start = start(region[recombOvs$subjectHits,])) 
  dist <- recombOvsDF %>% 
    mutate(mid = start + 5000) %>% 
    mutate(distance = pos - mid) %>% 
    ggplot(aes(x = distance)) +
    geom_density(adjust = 1/2) +
    ggtitle(title)
  return(dist)
}


ptestesRecomb <- recombDist(testesVCF[[1]], regions_list[[1]], title = "Testes") 
ptrioRecomb <- recombDist(triosVCF[[1]], regions_list[[1]], title = "Trios")
ptrioPatRecomb <- recombDist(triosPatVCF[[1]], regions_list[[1]], title = "Trios Pat")
ptrioMatRecomb <- recombDist(triosMatVCF[[1]], regions_list[[1]], title = "Trios Mat")

ptestesRecomb + ptrioRecomb + ptrioPatRecomb + ptrioMatRecomb 
ptestesRecomb2 + ptrioRecomb2 + ptrioPatRecomb2 + ptrioMatRecomb2
```


## Genome Segments
```{r}
load("segments_regions_list.RData")

# Find overlaps between mutations and genomic regions
segments_dist = genomic_distribution(VCFs, surveyed_list, segments_regions_list)
segments_depl = enrichment_depletion_test(segments_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
segments_depl$region = factor(segments_depl$region, levels = c("TSS","PF","E","WE","CTCF","T","R"))
pSegments <- plot_enrichment_depletion_prop(segments_depl)

png(paste0(plotsPath, 'segments.png'), height=5, width=10, units = "in", res = 300)
pSegments 
dev.off()
```

# Other Analysis
## Custom Rep Timing
```{r density, fig.height=5, fig.width=4}
repTimeH9es <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeFsuRepliChip/wgEncodeFsuRepliChipH9esWaveSignalRep1.bigWig", format="BigWig")
repTimeH9es <- keepStandardChromosomes(repTimeH9es, pruning.mode="coarse")
seqlevelsStyle(repTimeH9es) <- "NCBI"
export.bed(repTimeH9es, "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeH9es.bed")

repTimeGm12878 <- import("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqGm12878WaveSignalRep1.bigWig", format="BigWig")
seqlevelsStyle(repTimeGm12878) <- "NCBI"
export.bed(repTimeGm12878, "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeGm12878.bed")

repTimeLCL0 <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeLCL.txt", col_names = F, col_types = 'ccd') %>% 
  mutate(X3 = na_if(X3, "NaN")) %>% 
  mutate(X2 = if_else(str_detect(X2, "e"), "NA", X2)) %>% 
  mutate(X2 = as.numeric(na_if(X2, "NA"))) %>% 
  mutate(X1 = str_replace(X1, "23", "X")) %>% 
  mutate(X1 = str_replace(X1, "24", "Y")) %>% 
  drop_na() %>% 
  select(chrom = X1, start = X2, end = X2, score = X3)
repTimeLCL <- as(repTimeLCL0, "GRanges")
seqlevelsStyle(repTimeLCL) <- "NCBI"
export.bed(repTimeLCL, "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeLCL.bed")

#Create bed file with 1mb segments of genome
genome <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Surveyed_regions/Trio_DNMs.callable.bed", col_names = c("chr", "start", "end"), col_types = 'cii') %>% 
  mutate(rounded = round_any(end, 1000000, f = floor))
genome_1mb <- tibble()
for (i in genome$chr) {
  chrDF <- tibble(chr = i, start = seq(from=0, to = genome$rounded[genome$chr == i]-1000000, by = 1000000), end = seq(from = 999999, to = genome$rounded[genome$chr == i] -1, by = 1000000))
  genome_1mb <- bind_rows(genome_1mb, chrDF)
}
#Export for mapping to rep timing with bedtools map
write_tsv(genome_1mb, "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/genome_1mb.bed", col_names = F)

#Read back in
H9es1mb <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeH9es.1mb.bed", col_names = c("chr", "start", "end", "repTime"), col_types = 'ciic') %>% 
  mutate(repTime = if_else(repTime == ".", "NA", repTime)) %>% 
  mutate(repTime = as.numeric(repTime))
H9es1mb_gr <- import("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeH9es.1mb.bed")
seqlevelsStyle(H9es1mb_gr) <- "UCSC"
Gm128781mb <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeGm12878.1mb.bed", col_names = c("chr", "start", "end", "repTime"), col_types = 'ciic') %>% 
  mutate(repTime = if_else(repTime == ".", "NA", repTime)) %>% 
  mutate(repTime = as.numeric(repTime))
Gm128781mb_gr <- import("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeGm12878.1mb.bed")
seqlevelsStyle(Gm128781mb_gr) <- "UCSC"
LCL1mb <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeLCL.1mb.bed", col_names = c("chr", "start", "end", "repTime"), col_types = 'ciic') %>% 
  mutate(repTime = if_else(repTime == ".", "NA", repTime)) %>% 
  mutate(repTime = as.numeric(repTime))
LCL1mb_gr <- import("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/processTimingFiles/repTimeLCL.1mb.bed")
seqlevelsStyle(LCL1mb_gr) <- "UCSC"

#Get variant overlaps
repTimeDensity <- function(repTimeDF, repTimeGR) {
  testesOverlap <- findOverlaps(testesVCF[[1]], repTimeGR)
  testesRepTime <- tibble(chr = as.character(seqnames(testesVCF[[1]][testesOverlap@from])),
                          pos = start(testesVCF[[1]][testesOverlap@from]),
                          repTime = as.numeric(repTimeGR[testesOverlap@to]$name))
  
  colonOverlap <- findOverlaps(colonVCF[[1]], repTimeGR)
  colonRepTime <- tibble(chr = as.character(seqnames(colonVCF[[1]][colonOverlap@from])),
                          pos = start(colonVCF[[1]][colonOverlap@from]),
                          repTime = as.numeric(repTimeGR[colonOverlap@to]$name))
  
  triosOverlap <- findOverlaps(triosVCF[[1]], repTimeGR)
  triosRepTime <- tibble(chr = as.character(seqnames(triosVCF[[1]][triosOverlap@from])),
                          pos = start(triosVCF[[1]][triosOverlap@from]),
                          repTime = as.numeric(repTimeGR[triosOverlap@to]$name))
  
  triosPatOverlap <- findOverlaps(triosPatVCF[[1]], repTimeGR)
  triosPatRepTime <- tibble(chr = as.character(seqnames(triosPatVCF[[1]][triosPatOverlap@from])),
                          pos = start(triosPatVCF[[1]][triosPatOverlap@from]),
                          repTime = as.numeric(repTimeGR[triosPatOverlap@to]$name))
  
  densityDF <- repTimeDF %>% mutate(source = "Genome") %>% 
    bind_rows(testesRepTime %>% mutate(source = "Testes")) %>% 
    bind_rows(colonRepTime %>% mutate(source = "Colon")) %>% 
    bind_rows(triosRepTime %>% mutate(source = "Trios")) %>% 
    bind_rows(triosPatRepTime %>% mutate(source = "Trios_Pat"))
  return(densityDF)
}

densityH9es1 <- repTimeDensity(H9es1mb, H9es1mb_gr) %>% ggplot(aes(x=repTime, color = source)) +
  geom_density(adjust = 1/2) +
  theme_minimal() +
  scale_x_reverse() +
  xlab("<- Early             RepTime (H9es)           Late ->")

densityGm12878 <- repTimeDensity(Gm128781mb, Gm128781mb_gr) %>% ggplot(aes(x=repTime, color = source)) +
  geom_density(adjust = 1/2) +
  theme_minimal() +
  scale_x_reverse() +
  xlab("<- Early             RepTime (Gm12878)           Late ->")

densityLCL <- repTimeDensity(LCL1mb, LCL1mb_gr) %>% ggplot(aes(x=repTime, color = source)) +
  geom_density(adjust = 1/2) +
  theme_minimal() +
  scale_x_reverse() +
  xlab("<- Early             RepTime (LCL)           Late ->")

densityH9es1  / densityLCL / densityGm12878 + plot_layout(guides = 'collect')
```

## SigFit Dataset
```{r SigFitData}
repTimingFolder = "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/ReplicationTiming/SigFit/"
pyrTri_to_mutOpp <- c(rep(c(1:4,9:12,17:20,25:28), 3), rep(c(5:8,13:16,21:24,29:32),3))

#Use signatures from cosmic but edit 10b and add 10c to match signatures from POL mutant paper
data("cosmic_signatures_v3")
polSigs <- read.table("POLpaperSignatures.txt") %>% 
  select(SBS10B,SBS10C) %>% 
  t()
cosmic_signatures_v3_custom10 <- rbind(cosmic_signatures_v3, polSigs)


getMutTable <- function(mutFolder, suffix) {
  files <- list.files(paste0(repTimingFolder, mutFolder), pattern = suffix, full.names = TRUE)
  samples <-  str_remove(list.files(paste0(repTimingFolder, mutFolder), pattern = suffix), suffix)
  VCF <- read_vcfs_as_granges(files, samples, ref_genome, check_alleles=T)
  muts <- t(mut_matrix(vcf_list = VCF, ref_genome))
  return(muts)
}

getOppTable <- function(mutFolder, suffix) {
  pyrCountFiles <- list.files(paste0(repTimingFolder, mutFolder), pattern = suffix, full.names = TRUE)
  samples <-  str_remove(list.files(paste0(repTimingFolder, mutFolder), pattern = suffix), suffix)
  opps <- NULL
  for(i in pyrCountFiles) {
    pyrDF <- read_tsv(i, col_names = F, col_types = 'ci')[pyrTri_to_mutOpp,2]
    opps <- bind_cols(opps, pyrDF)
  }
  opps <- as_tibble(t(opps), rownames = NULL, colnames = NULL)
  return(opps)
}


#Default sigs 1, 5 and 18 for testes and colon
getExposureCustom <- function(mutFolder, suffix, adjust = F, header = "", sigChoice = c("SBS1", "SBS5", "SBS18")) {
  muts <- getMutTable(mutFolder, suffix = paste0(header, suffix, ".vcf"))
  if (adjust == F) {
    sigs <- cosmic_signatures_v3_custom10
  }
  else {
    opp_to <- getOppTable(mutFolder, suffix = paste0(suffix, ".tri_freqs.pyr.tsv"))  %>% 
  summarise_all(list(mean))
    toMat <- bind_rows(replicate(nrow(cosmic_signatures_v3_custom10), opp_to, simplify = FALSE))
    sigs <- convert_signatures(cosmic_signatures_v3_custom10, 
                                   opportunities_from = "human-genome",
                                   opportunities_to = toMat)
  }
  selected_signatures <- sigs[rownames(sigs) %in% sigChoice,]
  fit <- fit_signatures(counts = muts, signatures = selected_signatures,
                                   iter = 4000, 
                                   warmup = 2000, 
                                   chains = 1, 
                                   seed = 1756)
  exposures <- retrieve_pars(fit, par = "exposures") [[1]]
  return(exposures)
}


#Testes and colon. N = no adjustment for coverage/rep regions
ptestesEarlyN <- getExposureCustom(mutFolder = "Testes/", suffix = ".earlyRep")
ptestesEarly <- getExposureCustom(mutFolder = "Testes/", suffix = ".earlyRep", adjust = T)
ptestesIntermediateN <- getExposureCustom(mutFolder = "Testes/", suffix = ".intermediateRep")
ptestesIntermediate <- getExposureCustom(mutFolder = "Testes/", suffix = ".intermediateRep", adjust = T)
ptestesLateN <- getExposureCustom(mutFolder = "Testes/", suffix = ".lateRep")
ptestesLate <- getExposureCustom(mutFolder = "Testes/", suffix = ".lateRep", adjust = T)

pcolonEarlyN <- getExposureCustom(mutFolder = "Colon/", suffix = ".earlyRep")
pcolonEarly <- getExposureCustom(mutFolder = "Colon/", suffix = ".earlyRep", adjust = T)
pcolonIntermediateN <- getExposureCustom(mutFolder = "Colon/", suffix = ".intermediateRep")
pcolonIntermediate <- getExposureCustom(mutFolder = "Colon/", suffix = ".intermediateRep", adjust = T)
pcolonLateN <- getExposureCustom(mutFolder = "Colon/", suffix = ".lateRep")
pcolonLate <- getExposureCustom(mutFolder = "Colon/", suffix = ".lateRep", adjust = T)

#Lung never-smoker(normal) and exsmoker
plungNormalEarly <- getExposureCustom(mutFolder = "Lung_Normal/", suffix = ".earlyRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
plungNormalIntermediate <- getExposureCustom(mutFolder = "Lung_Normal/", suffix = ".intermediateRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
plungNormalLate <- getExposureCustom(mutFolder = "Lung_Normal/", suffix = ".lateRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
plungExsmokerEarly <- getExposureCustom(mutFolder = "Lung_ExSmoker/", suffix = ".earlyRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
plungExsmokerIntermediate <- getExposureCustom(mutFolder = "Lung_ExSmoker/", suffix = ".intermediateRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
plungExsmokerLate <- getExposureCustom(mutFolder = "Lung_ExSmoker/", suffix = ".lateRep", adjust = T, sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
#General Genomes
ptriosEarly <- getExposureCustom(mutFolder = "Genome/", suffix = ".earlyRep", adjust = T, header = "trios")
ptriosIntermediate <- getExposureCustom(mutFolder = "Genome/", suffix = ".intermediateRep", adjust = T, header = "trios")
ptriosLate <- getExposureCustom(mutFolder = "Genome/", suffix = ".lateRep", adjust = T, header = "trios")

ptriosPatEarly <- getExposureCustom(mutFolder = "Genome/", suffix = ".earlyRep", adjust = T, header = "trios_pat")
ptriosPatIntermediate <- getExposureCustom(mutFolder = "Genome/", suffix = ".intermediateRep", adjust = T, header = "trios_pat")
ptriosPatLate <- getExposureCustom(mutFolder = "Genome/", suffix = ".lateRep", adjust = T, header = "trios_pat")

ptriosMatEarly <- getExposureCustom(mutFolder = "Genome/", suffix = ".earlyRep", adjust = T, header = "trios_mat")
ptriosMatIntermediate <- getExposureCustom(mutFolder = "Genome/", suffix = ".intermediateRep", adjust = T, header = "trios_mat")
ptriosMatLate <- getExposureCustom(mutFolder = "Genome/", suffix = ".lateRep", adjust = T, header = "trios_mat")

#PolMutants
pPolEEarly <- getExposureCustom(mutFolder = "Sperm/", suffix = ".earlyRep", adjust = T, header = "22", sigChoice = c("SBS1","SBS5","SBS10a", "SBS10B", "SBS28"))
pPolEIntermediate <- getExposureCustom(mutFolder = "Sperm/", suffix = ".intermediateRep", adjust = T, header = "22", sigChoice = c("SBS1","SBS5","SBS10a", "SBS10B", "SBS28"))
pPolELate <- getExposureCustom(mutFolder = "Sperm/", suffix = ".lateRep", adjust = T, header = "22", sigChoice = c("SBS1","SBS5","SBS10a", "SBS10B", "SBS28"))

pPolDEarly <- getExposureCustom(mutFolder = "Sperm/", suffix = ".earlyRep", adjust = T, header = "23", sigChoice = c("SBS1","SBS5","SBS10C","SBS28"))
pPolDIntermediate <- getExposureCustom(mutFolder = "Sperm/", suffix = ".intermediateRep", adjust = T, header = "23", sigChoice = c("SBS1","SBS5","SBS10C","SBS28"))
pPolDLate <- getExposureCustom(mutFolder = "Sperm/", suffix = ".lateRep", adjust = T, header = "23", sigChoice = c("SBS1","SBS5","SBS10C","SBS28"))



```



## SigFit Plotting
```{r SigFit, fig.height=7, fig.width=10}
p1 <- function(exposures, timing = "samples") {
    exposurePlot <- tibble(sample = row.names(exposures), exposures) %>% 
      pivot_longer(-sample, names_to = "signature") %>% 
      mutate(signature = fct_relevel(signature, rev(sigChoice))) %>% 
      ggplot(aes(fill = signature, y = value, x = sample)) +
        geom_bar(position="stack", stat="identity") +
        ylab("Proportion") +
        xlab(timing) +
        theme_minimal() +
        #theme(axis.text.x = element_blank())
        theme(axis.text.x = element_text(angle = 90))
  return(exposurePlot)
}

#Plot for each sample
sigChoice <-  c("SBS1","SBS5", "SBS18")
(p1(ptestesEarlyN, timing = "Early") + p1(ptestesIntermediateN, timing = "Intermediate") + p1(ptestesLateN, timing = "Late"))/(p1(pcolonEarlyN, timing = "Early") + p1(pcolonIntermediateN, timing = "Intermediate") + p1(pcolonLateN, timing = "Late")) + plot_layout(guides = 'collect')
(p1(ptestesEarly, timing = "Early") + p1(ptestesIntermediate, timing = "Intermediate") + p1(ptestesLate, timing = "Late"))/(p1(pcolonEarly, timing = "Early") + p1(pcolonIntermediate, timing = "Intermediate") + p1(pcolonLate, timing = "Late")) + plot_layout(guides = 'collect')
#Pol sperm
(p1(pPolEEarly, timing = "Early") + p1(pPolEIntermediate, timing = "Intermediate") + p1(pPolELate, timing = "Late"))/(p1(pPolDEarly, timing = "Early") + p1(pPolDIntermediate, timing = "Intermediate") + p1(pPolDLate, timing = "Late")) + plot_layout(guides = 'collect')

# (p1(ptriosEarly, timing = "Early") + p1(ptriosIntermediate, timing = "Intermediate") + p1(ptriosLate, timing = "Late")) / 
#   (p1(ptriosPatEarly, timing = "Early") + p1(ptriosPatIntermediate, timing = "Intermediate") + p1(ptriosPatLate, timing = "Late")) /
#   (p1(ptriosMatEarly, timing = "Early") + p1(ptriosMatIntermediate, timing = "Intermediate") + p1(ptriosMatLate, timing = "Late")) +
#   plot_layout(guides = 'collect')
sigChoice <-  c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26")
(p1(plungNormalEarly, timing = "Early") + p1(plungNormalIntermediate, timing = "Intermediate") + p1(plungNormalLate, timing = "Late"))/(p1(plungExsmokerEarly, timing = "Early") + p1(plungExsmokerIntermediate, timing = "Intermediate") + p1(plungExsmokerLate, timing = "Late")) + plot_layout(guides = 'collect')

```

## SigFit Scaled
```{r, fig.height=6, fig.width=6}
pScaled <- function(early, int, late, exp = list(), obs = list(), tissue, sigChoice = c("SBS1", "SBS5", "SBS18"), colours = c("#F8766D", "#00BA38", "#619CFF")) {
  p0 <- early %>% mutate(timing = "Early") %>% 
    bind_rows(int %>% mutate(timing = "Intermediate")) %>% 
    bind_rows(late %>% mutate(timing = "Late")) %>% 
    group_by(timing) %>% 
    summarise_all(list(mean)) %>% 
    mutate(expected = exp) %>% 
    mutate(observed = obs) 
  p <- p0 %>% 
    #Convert to proportion of expected
    mutate_at(.vars = sigChoice, list(~(observed*.)/expected)) %>% 
    select(-expected, -observed) %>% 
    pivot_longer(-timing, names_to = "signature") %>% 
    mutate(signature = fct_relevel(signature, rev(sigChoice))) %>% 
    ggplot(aes(fill = signature, y = value, x = timing)) +
        geom_bar(colour = "black", position="stack", stat="identity") +
        #scale_fill_manual(values=colours) +
        ylim(0,1.5) +
        ylab("Proportion of Expected") +
        xlab(tissue) +
        theme_bw()
  return(p)
}
testesScaled <- pScaled(ptestesEarly, ptestesIntermediate, ptestesLate, exp = c(5360.486, 7549.530, 5421.352), obs = c(5609, 7439, 5329), tissue = "Testes")
colonScaled <- pScaled(pcolonEarly, pcolonIntermediate, pcolonLate, exp = c(6444.959, 9075.888, 7445.242), obs = c(4384, 7982, 9716), tissue = "Colon")

testesScaled + colonScaled + plot_layout(guides = 'collect')

triosScaled <- pScaled(ptriosEarly, ptriosIntermediate, ptriosLate, exp = c(57850.367, 81490.178, 72387.216), obs = c(64886, 87697, 63358), tissue = "Trios")
triosPatScaled <- pScaled(ptriosPatEarly, ptriosPatIntermediate, ptriosPatLate, exp = c(57850.367, 81490.178, 72387.216), obs = c(64886, 87697, 63358), tissue = "triosPat")
triosMatScaled <- pScaled(ptriosMatEarly, ptriosMatIntermediate, ptriosMatLate, exp = c(57850.367, 81490.178, 72387.216), obs = c(64886, 87697, 63358), tissue = "triosMat")

triosScaled + triosPatScaled + triosMatScaled + plot_layout(guides = 'collect')

lungNormalScaled <- pScaled(plungNormalEarly, plungNormalIntermediate, plungNormalLate, exp = c(60525.507, 85258.479, 75734.575), obs = c(53172, 86600, 86013), tissue = "lungNormal", sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
lungExsmokerScaled <- pScaled(plungExsmokerEarly, plungExsmokerIntermediate, plungExsmokerLate, exp = c(175233.711, 246840.722, 219267.072), obs = c(118218, 225768, 308480), tissue = "lungExsmoker", sigChoice = c("SBS1","SBS2","SBS4","SBS5","SBS12", "SBS13", "SBS26"))
lungNormalScaled + lungExsmokerScaled + plot_layout(guides = 'collect')

testesScaled + colonScaled + lungNormalScaled + lungExsmokerScaled + plot_layout(guides = 'collect')

#PolE and D
PolErep <- repTiming_dist %>% 
  filter(sample == "POLE_Sperm") 
PolDrep <- repTiming_dist %>% 
  filter(sample == "POLD1_Sperm") 
  
PolEScaled <- pScaled(pPolEEarly, pPolEIntermediate, pPolELate, exp = PolErep$expected, obs = PolErep$observed, tissue = "PolE Sperm", sigChoice =  c("SBS1","SBS5","SBS10a", "SBS10B", "SBS28"))
PolDScaled <- pScaled(pPolDEarly, pPolDIntermediate, pPolDLate, exp = PolDrep$expected, obs = PolDrep$observed, tissue = "PolD1 Sperm", sigChoice = c("SBS1","SBS5", "SBS10C", "SBS28"))

png('plots/repTimePolSperm.png', height=5, width=7, units = "in", res = 300)
PolEScaled + PolDScaled 
dev.off()

png('plots/repTimeSigsTestesColon.png', height=5, width=7, units = "in", res = 300)
colonScaled + testesScaled + plot_layout(guides = 'collect')
dev.off()

```


## Methylation Process Files
```{r}
methPath <- "/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/Methylation/split/"

#Testes RRBS
noMethTestis <- read_tsv("methylationTestis.bed", col_types = 'cccccccccccd') %>% 
  filter(percentMeth == 0) %>% 
  select(c(2:7))
write_tsv(noMethTestis, paste0(methPath, "testesNoMeth.bed"), col_names = F)
methTestis <- read_tsv("methylationTestis.bed", col_types = 'cccccccccccd') %>% 
  filter(percentMeth > 0)%>% 
  select(c(2:7))
write_tsv(methTestis, paste0(methPath, "testesMeth.bed"), col_names = F)

#Lung RRBS
noMethLung <- read_tsv("lungRRBS1.bed", col_types = 'cccccccccccd') %>% 
  filter(percentMeth == 0) %>% 
  select(c(2:7))
write_tsv(noMethLung, paste0(methPath, "lungNoMeth.bed"), col_names = F)
methLung <- read_tsv("lungRRBS1.bed", col_types = 'cccccccccccd') %>% 
  filter(percentMeth > 0)%>% 
  select(c(2:7))
write_tsv(methLung, paste0(methPath, "lungMeth.bed"), col_names = F)

#Colon WGBS
#awk '$10>2' ENCFF424XKF.bed > colonCov3.bed
#awk '$11>0' colonCov3.bed | cut -f 1-6 > colonMeth_38.bed
#awk '$11==0' colonCov3.bed | cut -f 1-6 > colonNoMeth_38.bed
#liftOver /lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/Methylation/colonMeth_38.bed hg38ToHg19.over.chain.gz /lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/colonMeth_37.bed /lustre/scratch119/casm/team294rr/rs30/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/colonMeth_38_unmapped.bed

# methColon <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/Methylation/split/colonMeth_19.bed", col_names = F, col_types = 'ciiccc')
# noMethColon <- read_tsv("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/Methylation/split/colonMeth_19.bed", col_names = F, col_types = 'ciiccc')

```

## Methylation Enrichment
```{r, fig.height=4, fig.width=4}
# regions bed files
meth_regions_files = list.files("/Users/mn7/volumes/rs30_lustre/Testicular_project/Final_Analyses/Unmatched_WGS/Signature_Analyses/Methylation/split", full.names = T)
print(meth_regions_files)
# 
# # region bed files
# meth_regions_list=list()
# for(i in meth_regions_files){
#   print(paste("Importing", i))
#   meth_region <- import(i)
#   ## Import the file using rtracklayer and use the UCSC naming standard
#   seqlevelsStyle(meth_region) <- "UCSC"
#   meth_regions_list[[i]] <- meth_region
# }
# 
# names(meth_regions_list) = c("colonMeth", "colonNoMeth", "lungMeth", "lungNoMeth","testesMeth", "testesNoMeth")
# save(meth_regions_list, file="meth_regions_list.RData")

load("meth_regions_list.RData")

meth_dist = genomic_distribution(VCFs, surveyed_list, meth_regions_list)
meth_depl = enrichment_depletion_test(meth_dist, by = tissues) %>% 
  #Multiple test correction
  mutate(significant = if_else(pval < p_corrected, "*", ""))
meth_depl$region = factor(meth_depl$region, levels = c("colonMeth", "colonNoMeth", "lungMeth", "lungNoMeth","testesMeth", "testesNoMeth"))
plot_enrichment_depletion_prop(meth_depl %>% filter(region %in%  c("colonMeth", "colonNoMeth")))
plot_enrichment_depletion_prop(meth_depl %>% filter(region %in%  c("lungMeth", "lungNoMeth")))
plot_enrichment_depletion_prop(meth_depl %>% filter(region %in%  c("testesMeth", "testesNoMeth")))

#Fraction with any methylation

methFraction0 <- meth_dist %>% 
  filter(str_detect(region, "colon") & sample == "Colon" | str_detect(region, "lung") & sample == "LungNormal" | str_detect(region, "testes") & sample == "Testes" ) %>% 
  mutate(methylation = c(T,F,T,F,T,F))
## Report plot just testes colon
# methFraction0 <- meth_dist %>% 
#   filter(str_detect(region, "colon") & sample == "Colon" | str_detect(region, "testes") & sample == "Testes" ) %>% 
#   mutate(methylation = c(T,F,T,F))
methTotals <- methFraction0 %>% 
  group_by(sample) %>% 
  summarise(tot = sum(surveyed_region_length))

methFraction <- methFraction0 %>% 
  left_join(methFraction0 %>% group_by(sample) %>% summarise(tot = sum(surveyed_region_length)), by = "sample") %>% 
  mutate(fraction = surveyed_region_length/tot) %>% 
  mutate(obsExp = observed/expected) %>% 
  mutate(meth = if_else(methylation, "TRUE", "FALSE")) %>% 
  mutate(meth = fct_relevel(meth, c("TRUE", "FALSE")))
  
    

pFraction <- ggplot(methFraction, aes(x = meth, y = fraction, fill = sample,  
        alpha = methylation)) + geom_bar(colour = "black", stat = "identity", 
        position = position_dodge()) + facet_grid(~sample) + 
        theme_bw() + theme(axis.ticks = element_blank(), axis.text.x = element_blank()) +
        xlab("") + ylab("CpG Fraction")  + 
        scale_x_discrete(breaks = NULL)


pObsExp <- ggplot(methFraction, aes(x = meth, y = obsExp, fill = sample,  
        alpha = methylation)) + geom_bar(colour = "black", stat = "identity", 
        position = position_dodge()) + facet_grid(~sample) + 
        theme_bw() + theme(axis.ticks = element_blank(), axis.text.x = element_blank()) +
        xlab("") + ylab("Observed/Expected")  + 
        scale_x_discrete(breaks = NULL)

pFraction / pObsExp + plot_layout(guides = 'collect') + plot_layout(heights = c(2, 1.2))

png('plots/methylationFraction.png', height=5, width=5, units = "in", res = 300)
pFraction / pObsExp + plot_layout(guides = 'collect') + plot_layout(heights = c(2, 1.2))
dev.off()
```



## Expression Model
```{r glmm, fig.height=3, fig.width=5}
get_gene_df <- function() {
  biomart <- read_tsv("/Volumes/GoogleDrive/My Drive/Testes/gene_prot_coords.tsv", col_types = 'ccciiiii') %>% 
    drop_na(prot) %>% 
    dplyr::rename(prot_id = prot, gene_id = gene)
  
  data("refcds_hg19", package="dndscv")
  gene_df <- tibble(gene_name = sapply(RefCDS, function(x) x$gene_name),
                    prot_id = sapply(RefCDS, function(x) x$protein_id)) %>% 
    left_join(biomart, by = "prot_id") %>% 
    mutate(length = end - start)
  gene_gr <- GRanges(gene_df$chr, IRanges(start=gene_df$start,end=gene_df$end), strand = gene_df$strand)
  seqlevelsStyle(gene_gr) <- "UCSC"
  gene_df <- gene_df %>% 
    mutate(gc_content = gcContentCalc(gene_gr, BSgenome.Hsapiens.UCSC.hg19)) %>% 
    select(gene_name, chr, start, end, strand, length, gc_content)
  return(gene_df)
}
gene_df <- get_gene_df() 
rm(RefCDS, gr_genes)
gene_gr <- GRanges(gene_df$chr, IRanges(start=gene_df$start,end=gene_df$end))
seqlevelsStyle(gene_gr) <- "UCSC"

ovs<- as.data.frame(findOverlaps(groupedVCFs,gene_gr))
overlapHits <- cbind(condensedSource[ovs$queryHits,],gene_df[ovs$subjectHits,]) %>% 
  group_by(gene_name, source) %>% 
  dplyr::summarise(count = n()) %>% 
  ungroup() 

mutsDF0 <- tibble()
for(tissue in tissues)(mutsDF0 <- bind_rows(mutsDF0, gene_df %>% mutate(source = tissue)))

#Expression Data transcriptional scanning paper
fixDatesTable <- tibble(gene_name = c("01-Dec","01-Sep","02-Sep","03-Mar","03-Sep","04-Mar","04-Sep","05-Mar","05-Sep","06-Mar","06-Sep","07-Mar","07-Sep","08-Mar","08-Sep","09-Mar","09-Sep","10-Mar","10-Sep","11-Mar","11-Sep","12-Sep","14-Sep"),
                        fix = c("DEC1","SEPT1","SEPT2","MARCH3","SEPT3","MARCH4","SEPT4","MARCH5","SEPT5","MARCH6","SEPT6","MARCH7","SEPT7","MARCH8","SEPT8","MARCH9","SEPT9","MARCH10","SEPT10","MARCH11","SEPT11","SEPT12","SEPT14"))

germCellExpr <- read_csv("/Volumes/GoogleDrive/My Drive/Testes/exprFiles/expr_level_groups.csv", col_types = cols(.default = "c")) %>% 
  pivot_longer(1:9, names_to = "group", values_to = "gene_name") %>% 
  drop_na() %>% 
  left_join(fixDatesTable, by ="gene_name") %>% 
  mutate(gene_name = if_else(gene_name %in% fixDatesTable$gene_name, fix, gene_name)) %>% 
  select(-fix) 

mutsDF <- mutsDF0 %>% 
  left_join(overlapHits, by = c("gene_name", "source")) %>% 
  mutate(count = replace_na(count, 0)) %>% 
  left_join(germCellExpr, by = "gene_name") %>% 
  mutate(group = fct_relevel(group, c("Group_Unexp","Group_1","Group_2","Group_3","Group_4","Group_5","Group_6","Group_7","Group_8"))) %>% 
  drop_na()

plot_glmm <- function(mutTable, dataSources, exprGroups, tests = 8, formula) {
  #Construct formula and run model
  df <- tibble()
  for (dataset in dataSources) {
    fit_full <- glmer(formula = as.formula(formula), data = (mutTable %>% filter(source == dataset)), family = poisson(),
                    control=glmerControl(optimizer="bobyqa", optCtrl=list(maxfun=2e5)))
    #print(dispersion_glmer(fit_full))
    #print(summary(fit_full, corr = F))
    tidy_fit_full <- tidy(fit_full) %>% 
      mutate(group = str_remove(term,"group")) %>% 
      mutate(p.format = signif(p.value, 2)) %>% 
      mutate(source = dataset)
    df <- df %>% bind_rows(tidy_fit_full[2:9,])
  }
  #Plot together
  sigThresh <- 0.05 / tests
  df2 <- df %>% 
    mutate(p.format = if_else(p.value < sigThresh, "*", ""))
  df2$group <- factor(df2$group, levels = exprGroups)
  df2$source <- factor(df2$source, levels = dataSources)
  plot <- ggplot(df2, aes(x = group, y = estimate, colour = source, label = p.format, group = interaction(group, source))) +
    geom_hline(yintercept = 0, colour = "grey") +
    geom_point(position = position_dodge(0.9), size = 3) +
    geom_errorbar(aes(ymin=estimate-(std.error*1.96), ymax=estimate+(std.error*1.96)), width=.1, position = position_dodge(0.9)) +
    geom_text(aes(y = estimate+(std.error*1.96)), colour = "black", size = 4, position = position_dodge(0.9), vjust = -0.75) +
    theme_minimal() +
    xlab("Expression bin") + ylab("Effect estimate") + labs(colour = "Variant Source")
  return(plot)
}
#Plot for individual datasets or all together
glmmTissues <- c("Testes", "Trio_DNMs")
#glmmTissues <- tissues
glmmPlot <- plot_glmm(mutTable = mutsDF, dataSources = glmmTissues, exprGroups = c("Group_1","Group_2","Group_3","Group_4","Group_5","Group_6","Group_7","Group_8"), tests = 8, formula = "count ~ group + offset(log(length)) + offset(log(gc_content)) + (1|gene_name)")
glmmPlot

pdf('plots/glmmExpression.pdf', height=5, width=10)
glmmPlot
dev.off()

png('plots/glmmExpression.png', height=5, width=10, units = "in", res = 300)
glmmPlot
dev.off()
```



#RemovedDatasets/Code
```{r oldCode}

```
